{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import numpy as np \n",
    "import pandas as pd \n",
    "from Bio import Align \n",
    "import os \n",
    "from selenobot.tools import CDHIT, MMseqs\n",
    "from selenobot.datasets import *\n",
    "\n",
    "%load_ext autoreload \n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hidden Markov Model baseline\n",
    "\n",
    "I think it is important to have an HMM baseline against which to compare selenobot performance. To use an HMM as a classifier, I will need at least one HMM per category (i.e. full-length, truncated selenoproteins, and truncated non-selenoproteins); I suspect I will need more granularity than this, as HMMs are built using multi-sequence alignments, which require some degree of sequence similarity. \n",
    "\n",
    "## Training\n",
    "\n",
    " For building HMMs, I am thinking of doing the following:\n",
    "\n",
    "1. Cluster sequences in each category from the training set at 50 percent similarity using CD-HIT. \n",
    "2. Generate a multi-sequence alignment (MSA) for each category (possibly just using BioPython?)\n",
    "3. Use PyHMMER (a Python wrapper for HMMER3) to build an HMM for each cluster. \n",
    "\n",
    "The result of training will be a map from each overarching category (full-length, truncated selenoproteins, and truncated non-selenoproteins) to a set of HMMs\n",
    "\n",
    "## Prediction\n",
    "\n",
    "Generate a score for each sequence, for each generated HMM. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_dataset = TernaryDataset.from_hdf('../data/train.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDHIT._run: Using pre-saved clustering results at /home/prichter/Documents/selenobot/notebooks/cluster_hmm_0\n",
      "easy-cluster /home/prichter/Documents/selenobot/notebooks/hmm_0.fa hmm_0 tmp --min-seq-id 0.2 \n",
      "\n",
      "MMseqs Version:                     \t16.747c6\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "Target search mode                  \t0\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t16\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.2\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Weight file name                    \t\n",
      "Cluster Weight threshold            \t0.9\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \ttrue\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "Database type                       \t0\n",
      "Shuffle input database              \ttrue\n",
      "Createdb mode                       \t1\n",
      "Write lookup file                   \t0\n",
      "Offset of numeric ids               \t0\n",
      "\n",
      "createdb /home/prichter/Documents/selenobot/notebooks/hmm_0.fa tmp/17520026234266495660/input --dbtype 0 --shuffle 1 --createdb-mode 1 --write-lookup 0 --id-offset 0 --compressed 0 -v 3 \n",
      "\n",
      "Shuffle database cannot be combined with --createdb-mode 0\n",
      "We recompute with --shuffle 0\n",
      "Converting sequences\n",
      "[Multiline fasta can not be combined with --createdb-mode 0\n",
      "We recompute with --createdb-mode 1\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 0ms\n",
      "[==========\n",
      "Time for merging to input_h: 0h 0m 0s 0ms\n",
      "Time for merging to input: 0h 0m 0s 6ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 143ms\n",
      "Create directory tmp/17520026234266495660/clu_tmp\n",
      "cluster tmp/17520026234266495660/input tmp/17520026234266495660/clu tmp/17520026234266495660/clu_tmp --max-seqs 20 -c 0.8 --spaced-kmer-mode 1 --alignment-mode 3 -e 0.001 --min-seq-id 0.2 --remove-tmp-files 1 \n",
      "\n",
      "Set cluster sensitivity to -s 6.000000\n",
      "Set cluster mode SET COVER\n",
      "Set cluster iterations to 3\n",
      "linclust tmp/17520026234266495660/input tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 16 --compressed 0 -v 3 --cluster-weight-threshold 0.9 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.2 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 1 --force-reuse 0 \n",
      "\n",
      "kmermatcher tmp/17520026234266495660/input tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.2 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 16 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "kmermatcher tmp/17520026234266495660/input tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0.2 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 16 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Database size: 109036 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 109.04K 0s 227ms\n",
      "Sort kmer 0h 0m 0s 64ms\n",
      "Sort by rep. sequence 0h 0m 0s 16ms\n",
      "Time for fill: 0h 0m 0s 25ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 393ms\n",
      "rescorediagonal tmp/17520026234266495660/input tmp/17520026234266495660/input tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 109.04K 0s 52ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 37ms\n",
      "Time for processing: 0h 0m 0s 134ms\n",
      "clust tmp/17520026234266495660/input tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_rescore1 tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pre_clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 16 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 109.04K 0s 20ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 128764 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 109.04K 0s 20ms\n",
      "Add missing connections\n",
      "[=================================================================] 109.04K 0s 12ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 74ms\n",
      "Total time: 0h 0m 0s 96ms\n",
      "\n",
      "Size of the sequence database: 109036\n",
      "Size of the alignment database: 109036\n",
      "Number of clusters: 60131\n",
      "\n",
      "Writing results 0h 0m 0s 15ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 147ms\n",
      "createsubdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/order_redundancy tmp/17520026234266495660/input tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 27ms\n",
      "createsubdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/order_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 35ms\n",
      "filterdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_filter1 tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_filter2 --filter-file tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/order_redundancy --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 60.13K 0s 16ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 16ms\n",
      "Time for processing: 0h 0m 0s 66ms\n",
      "rescorediagonal tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_filter2 tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.2 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Can not find any score per column for coverage 0.800000 and sequence identity 0.200000. No hit will be filtered.\n",
      "[=================================================================] 60.13K 0s 30ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 23ms\n",
      "Time for processing: 0h 0m 0s 82ms\n",
      "align tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_rescore2 tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.2 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 60131 type: Aminoacid\n",
      "Target database size: 60131 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 60.13K 4s 146ms\n",
      "Time for merging to aln: 0h 0m 0s 16ms\n",
      "106058 alignments calculated\n",
      "101810 sequence pairs passed the thresholds (0.959946 of overall calculated)\n",
      "1.693137 hits per query sequence\n",
      "Time for processing: 0h 0m 4s 192ms\n",
      "clust tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/aln tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/clust --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 16 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 60.13K 0s 8ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 41679 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 60.13K 0s 11ms\n",
      "Add missing connections\n",
      "[=================================================================] 60.13K 0s 5ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 39ms\n",
      "Total time: 0h 0m 0s 50ms\n",
      "\n",
      "Size of the sequence database: 60131\n",
      "Size of the alignment database: 60131\n",
      "Number of clusters: 39699\n",
      "\n",
      "Writing results 0h 0m 0s 11ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 82ms\n",
      "mergeclusters tmp/17520026234266495660/input tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pre_clust tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/clust --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 60.13K 0s 8ms\n",
      "Clustering step 2\n",
      "[=================================================================] 39.70K 0s 24ms\n",
      "Write merged clustering\n",
      "[=================================================================] 109.04K 0s 39ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 12ms\n",
      "Time for processing: 0h 0m 0s 77ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_filter1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_rescore1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pre_clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_filter2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/pref_rescore2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/linclust/18129914891694722864/clust -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_redundancy tmp/17520026234266495660/input tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "prefilter tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 39699 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 39699 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 39.70K 0s 189ms\n",
      "Index table: Masked residues: 0\n",
      "Index table: fill\n",
      "[=================================================================] 39.70K 0s 146ms\n",
      "Index statistics\n",
      "Entries:          5163835\n",
      "DB size:          517 MB\n",
      "Avg k-mer size:   0.080685\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t142\n",
      "    VTMHHG\t73\n",
      "    GGTHAY\t65\n",
      "    VADGPQ\t61\n",
      "    MGVHTS\t59\n",
      "    LDPHFT\t52\n",
      "    TPHAAM\t50\n",
      "    IMHDKT\t50\n",
      "    PNGSFN\t44\n",
      "    PWKRKR\t44\n",
      "Time for index table init: 0h 0m 0s 800ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 39699\n",
      "Target db start 1 to 39699\n",
      "[=================================================================] 39.70K 0s 399ms\n",
      "\n",
      "1.519923 k-mers per position\n",
      "232 DB matches per sequence\n",
      "0 overflows\n",
      "6 sequences passed prefiltering per query sequence\n",
      "4 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 15ms\n",
      "Time for processing: 0h 0m 1s 653ms\n",
      "align tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step0 tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.2 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 39699 type: Aminoacid\n",
      "Target database size: 39699 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 39.70K 10s 392ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 12ms\n",
      "243940 alignments calculated\n",
      "222715 sequence pairs passed the thresholds (0.912991 of overall calculated)\n",
      "5.610091 hits per query sequence\n",
      "Time for processing: 0h 0m 10s 444ms\n",
      "clust tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step0 tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step0 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 16 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 39.70K 0s 6ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 17408 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 39.70K 0s 9ms\n",
      "Add missing connections\n",
      "[=================================================================] 39.70K 0s 8ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 33ms\n",
      "Total time: 0h 0m 0s 40ms\n",
      "\n",
      "Size of the sequence database: 39699\n",
      "Size of the alignment database: 39699\n",
      "Number of clusters: 20398\n",
      "\n",
      "Writing results 0h 0m 0s 5ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 58ms\n",
      "createsubdb tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step0 tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "prefilter tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 3.5 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 20398 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 20398 type: Aminoacid\n",
      "Index table k-mer threshold: 131 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 20.40K 0s 125ms\n",
      "Index table: Masked residues: 0\n",
      "Index table: fill\n",
      "[=================================================================] 20.40K 0s 146ms\n",
      "Index statistics\n",
      "Entries:          5504939\n",
      "DB size:          519 MB\n",
      "Avg k-mer size:   0.086015\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t44\n",
      "    EVSGRP\t27\n",
      "    GQQVAR\t27\n",
      "    LESPDR\t27\n",
      "    ALGNGK\t18\n",
      "    LLDPAL\t18\n",
      "    GPGGSL\t18\n",
      "    GEGGTL\t18\n",
      "    AVRRRR\t18\n",
      "    AARAAG\t15\n",
      "Time for index table init: 0h 0m 0s 728ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 131\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 20398\n",
      "Target db start 1 to 20398\n",
      "[=================================================================] 20.40K 1s 176ms\n",
      "\n",
      "32.866569 k-mers per position\n",
      "982 DB matches per sequence\n",
      "0 overflows\n",
      "7 sequences passed prefiltering per query sequence\n",
      "6 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 6ms\n",
      "Time for processing: 0h 0m 2s 392ms\n",
      "align tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.2 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 20398 type: Aminoacid\n",
      "Target database size: 20398 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 20.40K 1s 46ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 6ms\n",
      "78776 alignments calculated\n",
      "45748 sequence pairs passed the thresholds (0.580735 of overall calculated)\n",
      "2.242769 hits per query sequence\n",
      "Time for processing: 0h 0m 1s 70ms\n",
      "clust tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step1 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 16 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 20.40K 0s 3ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 3772 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 20.40K 0s 3ms\n",
      "Add missing connections\n",
      "[=================================================================] 20.40K 0s 1ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 13ms\n",
      "Total time: 0h 0m 0s 17ms\n",
      "\n",
      "Size of the sequence database: 20398\n",
      "Size of the alignment database: 20398\n",
      "Number of clusters: 14435\n",
      "\n",
      "Writing results 0h 0m 0s 2ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 29ms\n",
      "createsubdb tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "prefilter tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step2 tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step2 tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 6 -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 14435 type: Aminoacid\n",
      "Estimated memory consumption: 1018M\n",
      "Target database size: 14435 type: Aminoacid\n",
      "Index table k-mer threshold: 109 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 14.44K 0s 88ms\n",
      "Index table: Masked residues: 0\n",
      "Index table: fill\n",
      "[=================================================================] 14.44K 0s 110ms\n",
      "Index statistics\n",
      "Entries:          4128016\n",
      "DB size:          511 MB\n",
      "Avg k-mer size:   0.064500\n",
      "Top 10 k-mers\n",
      "    GPGGTL\t23\n",
      "    YTGTPK\t19\n",
      "    AALAAA\t15\n",
      "    AAAAGA\t15\n",
      "    ALAALA\t15\n",
      "    AAALLA\t15\n",
      "    LAAAAG\t15\n",
      "    TSSGKG\t15\n",
      "    LIGTGK\t15\n",
      "    AAAAAL\t15\n",
      "Time for index table init: 0h 0m 0s 686ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 109\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 14435\n",
      "Target db start 1 to 14435\n",
      "[=================================================================] 14.44K 6s 543ms\n",
      "\n",
      "380.257415 k-mers per position\n",
      "7786 DB matches per sequence\n",
      "0 overflows\n",
      "18 sequences passed prefiltering per query sequence\n",
      "20 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 7s 693ms\n",
      "align tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step2 tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step2 tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step2 tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.2 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 14435 type: Aminoacid\n",
      "Target database size: 14435 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 14.44K 0s 539ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 4ms\n",
      "74773 alignments calculated\n",
      "19084 sequence pairs passed the thresholds (0.255226 of overall calculated)\n",
      "1.322064 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 558ms\n",
      "clust tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step2 tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step2 tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step2 --cluster-mode 0 --max-iterations 1000 --similarity-type 2 --threads 16 --compressed 0 -v 3 --cluster-weight-threshold 0.9 \n",
      "\n",
      "Clustering mode: Set Cover\n",
      "[=================================================================] 14.44K 0s 1ms\n",
      "Sort entries\n",
      "Find missing connections\n",
      "Found 935 new connections.\n",
      "Reconstruct initial order\n",
      "[=================================================================] 14.44K 0s 1ms\n",
      "Add missing connections\n",
      "[=================================================================] 14.44K 0s 0ms\n",
      "\n",
      "Time for read in: 0h 0m 0s 8ms\n",
      "Total time: 0h 0m 0s 10ms\n",
      "\n",
      "Size of the sequence database: 14435\n",
      "Size of the alignment database: 14435\n",
      "Number of clusters: 12555\n",
      "\n",
      "Writing results 0h 0m 0s 1ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "mergeclusters tmp/17520026234266495660/input tmp/17520026234266495660/clu tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_redundancy tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step0 tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step1 tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step2 --threads 16 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 39.70K 0s 5ms\n",
      "Clustering step 2\n",
      "[=================================================================] 20.40K 0s 13ms\n",
      "Clustering step 3\n",
      "[=================================================================] 14.44K 0s 18ms\n",
      "Clustering step 4\n",
      "[=================================================================] 12.56K 0s 22ms\n",
      "Write merged clustering\n",
      "[=================================================================] 109.04K 0s 36ms\n",
      "Time for merging to clu: 0h 0m 0s 6ms\n",
      "Time for processing: 0h 0m 0s 63ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step_redundancy_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/pref_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/aln_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/clu_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step1_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step2 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_tmp/17287803400908755388/input_step2_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createtsv tmp/17520026234266495660/input tmp/17520026234266495660/input tmp/17520026234266495660/clu tmp/17520026234266495660/cluster.tsv --threads 16 -v 3 \n",
      "\n",
      "Time for merging to cluster.tsv: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 31ms\n",
      "result2repseq tmp/17520026234266495660/input tmp/17520026234266495660/clu tmp/17520026234266495660/clu_rep --db-load-mode 0 --compressed 0 --threads 16 -v 3 \n",
      "\n",
      "[=================================================================] 12.56K 0s 2ms\n",
      "Time for merging to clu_rep: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 29ms\n",
      "result2flat tmp/17520026234266495660/input tmp/17520026234266495660/input tmp/17520026234266495660/clu_rep tmp/17520026234266495660/rep_seq.fasta --use-fasta-header -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 30ms\n",
      "createseqfiledb tmp/17520026234266495660/input tmp/17520026234266495660/clu tmp/17520026234266495660/clu_seqs --threads 16 -v 3 \n",
      "\n",
      "[=================================================================] 12.56K 0s 12ms\n",
      "Time for merging to clu_seqs: 0h 0m 0s 4ms\n",
      "Time for processing: 0h 0m 0s 61ms\n",
      "result2flat tmp/17520026234266495660/input tmp/17520026234266495660/input tmp/17520026234266495660/clu_seqs tmp/17520026234266495660/all_seqs.fasta -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 108ms\n",
      "rmdb tmp/17520026234266495660/input -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb tmp/17520026234266495660/input_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu_seqs -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 7ms\n",
      "rmdb tmp/17520026234266495660/clu_rep -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "rmdb tmp/17520026234266495660/clu -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    }
   ],
   "source": [
    "for label, df in train_dataset.metadata.groupby('label'):\n",
    "    cdhit = CDHIT(df, c_cluster=0.5, cwd=os.getcwd(), name=f'hmm_{label}')\n",
    "    cdhit_df = cdhit.run(dereplicate=False, overwrite=False)\n",
    "    mmseqs = MMseqs(df, name=f'hmm_{label}', cwd=os.getcwd())\n",
    "    mmseqs_df = mmseqs.run(overwrite=True)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cdhit_cluster\n",
       "22490    4\n",
       "6301     4\n",
       "25953    4\n",
       "25937    4\n",
       "20062    4\n",
       "        ..\n",
       "16       1\n",
       "15       1\n",
       "14       1\n",
       "11       1\n",
       "8        1\n",
       "Length: 31112, dtype: int64"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_sizes = cdhit_df.groupby('cdhit_cluster').apply(len, include_groups=False).sort_values(ascending=False)\n",
    "cluster_sizes[cluster_sizes < 5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35948"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_sizes = mmseqs_df.groupby('mmseqs_cluster').apply(len, include_groups=False).sort_values(ascending=False)\n",
    "cluster_sizes[cluster_sizes < 5]\n",
    "len(mmseqs_df.mmseqs_cluster.unique())\n",
    "len(cdhit_df.cdhit_cluster.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "\n",
    "1. What is a good rule-of-thumb number for sequence similarity when generating multi-sequence alignments?\n",
    "2. Why use alignments instead of raw sequences when training HMMs?\n",
    "3. Should I de-replicate the sequences before building HMMs?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "selenobot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
