{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "import subprocess\n",
    "import math\n",
    "from tqdm import tqdm \n",
    "\n",
    "import dataframe_image as dfi \n",
    "\n",
    "import matplotlib.pyplot as plt \n",
    "from matplotlib.patches import Patch\n",
    "import matplotlib.cm \n",
    "import matplotlib.colors \n",
    "plt.rcParams['font.family'] = 'Arial'\n",
    "\n",
    "from selenobot.files import FASTAFile, fasta_file_parser_gtdb, GBFFFile, BLASTFile\n",
    "from selenobot.tools import BLAST\n",
    "from selenobot.genes import Gene, Genome\n",
    "\n",
    "%load_ext autoreload \n",
    "%autoreload 2\n",
    "\n",
    "model_organisms = dict()\n",
    "model_organisms['Pseudomonas aeruginosa'] = 'paer'\n",
    "model_organisms['Escherichia coli'] = 'ecol'\n",
    "model_organisms['Mycobacterium tuberculosis'] = 'mtub' \n",
    "model_organisms['Bacillus subtilis'] = 'bsub'\n",
    "model_organisms['Aliivibrio fischeri'] = 'afis'\n",
    "\n",
    "genome_metadata_df = pd.read_csv('../data/gtdb_genome_metadata_bacteria.csv', index_col=0)\n",
    "genome_metadata_df = genome_metadata_df[genome_metadata_df.species.isin(model_organisms.keys())]\n",
    "genome_metadata_df['code_name'] = genome_metadata_df.species.replace(model_organisms)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_summary_table(path='../data/model_organisms/summary_table.png'):   \n",
    "    table_df = list()\n",
    "    columns = ['ref. genome size', 'GTDB genome size', 'aligned sequences']\n",
    "    for species, code_name in model_organisms.items():\n",
    "        row = dict()\n",
    "        row['species'] = species\n",
    "\n",
    "        ref_fasta_file = FASTAFile(f'../data/model_organisms/ncbi_{code_name}_protein.faa')\n",
    "        gtdb_fasta_file = FASTAFile(f'../data/model_organisms/gtdb_{code_name}_protein.faa')\n",
    "        results_df = pd.read_csv(f'../data/model_organisms/{code_name}_results.csv', index_col=0)\n",
    "        row['ref. genome size'] = len(ref_fasta_file)\n",
    "        row['GTDB genome size'] = len(gtdb_fasta_file)\n",
    "        row['aligned sequences'] = results_df.blast_possible_match.sum()\n",
    "        table_df.append(row)\n",
    "    table_df = pd.DataFrame(table_df).set_index('species')\n",
    "    dfi.export(table_df, path)\n",
    "\n",
    "make_summary_table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_ncbi_data():\n",
    "    '''Download the genomes, proteins, and GBFF files from NCBI for each model organism.'''\n",
    "    genome_ids, code_names = list(genome_metadata_df.index), list(genome_metadata_df.code_name)\n",
    "\n",
    "    cmd = 'datasets download genome accession {genome_ids} --filename ../data/model_organisms/ncbi.zip --include protein,gbff,genome'.format(genome_ids=' '.join(genome_ids))\n",
    "    # subprocess.run(cmd, shell=True, check=True)\n",
    "    # subprocess.run('unzip ../data/model_organisms/ncbi.zip', shell=True, check=True)\n",
    "        \n",
    "    for code_name, genome_id in zip(code_names, genome_ids):\n",
    "        subprocess.run(f'cp ../data/model_organisms/ncbi/ncbi_dataset/data/{genome_id}/protein.faa ../data/model_organisms/ncbi_{code_name}_protein.faa', shell=True, check=True)\n",
    "        subprocess.run(f'cp ../data/model_organisms/ncbi/ncbi_dataset/data/{genome_id}/genomic.gbff ../data/model_organisms/ncbi_{code_name}_genomic.gbff', shell=True, check=True)\n",
    "        subprocess.run(f'cp ../data/model_organisms/ncbi/ncbi_dataset/data/{genome_id}/*_genomic.fna ../data/model_organisms/ncbi_{code_name}_genomic.fna', shell=True, check=True)\n",
    "\n",
    "\n",
    "def run_blast():\n",
    "    '''Run BLAST for the GTDB genome against the corresponding NCBI reference genome for each model organism.'''\n",
    "    blast = BLAST()\n",
    "\n",
    "    for species, code_name in model_organisms.items():\n",
    "        query_path = f'../data/model_organisms/gtdb_{code_name}_protein.faa'\n",
    "        subject_path = f'../data/model_organisms/ncbi_{code_name}_protein.faa'\n",
    "        output_path = f'../data/model_organisms/gtdb_{code_name}_protein.blast.tsv'\n",
    "        if not os.path.exists(output_path):\n",
    "            print(f'Running BLAST for {species}')\n",
    "            blast.run(query_path, subject_path, output_path=output_path, max_high_scoring_pairs=1, max_subject_sequences=1, make_database=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_cols = ['subject_id', 'sequence_identity', 'query_alignment_start', 'subject_alignment_start', 'query_sequence_length', 'subject_sequence_length', 'query_id']\n",
    "gtdb_fasta_cols = ['start', 'stop', 'strand', 'partial', 'seq']\n",
    "gbff_cols = ['gene', 'partial', 'strand', 'start', 'stop', 'copy_number', 'protein_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of NCBI proteins: 4607\n",
      "Number of GTDB proteins: 4717\n"
     ]
    }
   ],
   "source": [
    "ncbi_fasta_df = FASTAFile('../data/model_organisms/ncbi_ecol_protein.faa').to_df()\n",
    "gtdb_fasta_df = FASTAFile('../data/model_organisms/gtdb_ecol_protein.faa').to_df(parser=fasta_file_parser_gtdb)[gtdb_fasta_cols]\n",
    "\n",
    "print('\\nNumber of NCBI proteins:', len(ncbi_fasta_df))\n",
    "print('Number of GTDB proteins:', len(gtdb_fasta_df))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "find_matches: Found 4310 sequences in the input genome which exactly match sequences in the reference.\n",
      "find_boundary_errors: Found 193 sequences in the input genome with boundary errors.\n",
      "GBFFFile.__init__: Found 4818 coding sequences in the GBFF file.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GBFFFile.__init__: Parsing GBFF file entries.: 100%|██████████| 4818/4818 [00:00<00:00, 11063.47it/s]\n",
      "find_pseudogenes: Looking for overlapping pseudogenes in the reference genome...: 100%|██████████| 214/214 [00:00<00:00, 233.98it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "find_pseudogenes: Found 123 sequences in the input genome which match pseudogenes in the reference.\n",
      "GBFFFile.__init__: Found 4818 coding sequences in the GBFF file.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GBFFFile.__init__: Parsing GBFF file entries.: 100%|██████████| 4818/4818 [00:00<00:00, 10890.97it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GBFFFile.to_df: Removing 199 pseudogenes from the GBFF file.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "find_intergenic: Looking for intergenic regions in the reference genome...: 100%|██████████| 91/91 [00:00<00:00, 188.10it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "find_intergenic: Found 90 sequences in the input genome which are in intergenic regions.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Possible scenarios:\n",
    "# (1) Sequence is a pseudogene. \n",
    "# (2) Sequence lies in an intergenic region. \n",
    "# (3) Sequence overlaps with an annotated protein-coding gene in the reference genome (error). \n",
    "# (5) Sequence is in the right frame, but the length is wrong (boundary error). \n",
    "# (6) Sequence is correct. \n",
    "\n",
    "# Also probably important to account for the size of the overlap. \n",
    "\n",
    "def search(df:pd.DataFrame, start:int=None, stop:int=None, strand:int=None, top_hit:bool=True):\n",
    "    '''Look for sequences in the input GBFF file which overlap somewhere in the specified range.''' \n",
    "    def overlap(row):\n",
    "        return len(np.intersect1d(np.arange(start, stop), np.arange(row.start, row.stop)))\n",
    "    \n",
    "    if (strand is not None):\n",
    "        df = df[df.strand == strand]\n",
    "\n",
    "    df = df.copy()\n",
    "    df['overlap'] = df.apply(overlap, axis=1)\n",
    "    df['percent_overlap'] = 100 * (df.overlap / (stop - start))\n",
    "    df = df[df['overlap'] > 0]\n",
    "    df = df.sort_values('overlap', ascending=False)\n",
    "    return None if (len(df) == 0) else df.to_dict(orient='records')[0]\n",
    "\n",
    "\n",
    "def find_matches(df:pd.DataFrame, code_name:str='ecol'):\n",
    "    blast_df = BLASTFile(f'../data/model_organisms/gtdb_{code_name}_protein.blast.tsv').to_df()\n",
    "\n",
    "    def is_match(hit) -> bool:\n",
    "        if pd.isnull(hit.subject_id):\n",
    "            return False\n",
    "        if hit.sequence_identity < 95:\n",
    "            return False  \n",
    "        if not math.isclose(hit.subject_sequence_length, hit.query_sequence_length, abs_tol=5):\n",
    "            return False\n",
    "        return True\n",
    "\n",
    "    df, blast_df = df.align(blast_df, axis=0, join='left')\n",
    "    mask = blast_df.apply(is_match, axis=1)\n",
    "    \n",
    "    print(f'find_matches: Found {np.sum(mask)} sequences in the input genome which exactly match sequences in the reference.')\n",
    "    return df[mask], df[~mask], blast_df[mask]\n",
    "\n",
    "\n",
    "def find_boundary_errors(df:pd.DataFrame, code_name:str='ecol'):\n",
    "    ''''''\n",
    "    # NOTE: Seems worth distinguishing between left and right-side boundary errors. \n",
    "    blast_df = BLASTFile(f'../data/model_organisms/gtdb_{code_name}_protein.blast.tsv').to_df()\n",
    "\n",
    "    def is_boundary_error(hit) -> bool:\n",
    "        if pd.isnull(hit.subject_id):\n",
    "            return False\n",
    "        if hit.sequence_identity < 95:\n",
    "            return False\n",
    "        return True\n",
    "\n",
    "    df, blast_df = df.align(blast_df, axis=0, join='left')\n",
    "    mask = blast_df.apply(is_boundary_error, axis=1)\n",
    "\n",
    "    info_df = blast_df.copy[mask]\n",
    "    info_df['left_aligned'] = (info_df.query_alignment_start == 1) & (info_df.subject_alignment_start == 1)\n",
    "    info_df['right_aligned'] = boundary_errors_info_df.query_alignment_end == boundary_errors_info_df.query_sequence_length - 1) & (boundary_errors_info_df.subject_alignment_end == boundary_errors_info_df.subject_sequence_length)\n",
    "\n",
    "    print(f'find_boundary_errors: Found {np.sum(mask)} sequences in the input genome with boundary errors.')\n",
    "    return df[mask], df[~mask], blast_df[mask] \n",
    "\n",
    "\n",
    "def find_pseudogenes(df:pd.DataFrame, code_name:str='ecol'):\n",
    "    ''''''\n",
    "    gbff_df = GBFFFile(f'../data/model_organisms/ncbi_{code_name}_genomic.gbff').to_df() \n",
    "    gbff_df = gbff_df[gbff_df.pseudo].drop(columns=['protein_id', 'seq'])\n",
    "\n",
    "    def is_psuedogene(hit:dict, start:int=None, stop:int=None):\n",
    "        if hit is None:\n",
    "            return False, dict()\n",
    "        if (hit['start'] == start) or (hit['stop'] == stop):\n",
    "            return True, hit\n",
    "        if hit['percent_overlap'] > 95:\n",
    "            return True, hit\n",
    "        return False, dict()\n",
    "\n",
    "    info_df, mask = list(), list()\n",
    "    for row in tqdm(list(df.itertuples()), desc='find_pseudogenes: Looking for overlapping pseudogenes in the reference genome...'):\n",
    "        hit = search(gbff_df, start=int(row.start), stop=int(row.stop), strand=int(row.strand))\n",
    "        val, hit = is_psuedogene(hit, start=row.start, stop=row.stop)\n",
    "        hit['id'] = row.Index\n",
    "        info_df.append(hit)\n",
    "        mask.append(val)\n",
    "    info_df = pd.DataFrame(info_df).set_index('id')\n",
    "    mask = np.array(mask)\n",
    "\n",
    "    print(f'find_pseudogenes: Found {np.sum(mask)} sequences in the input genome which match pseudogenes in the reference.')\n",
    "    return df[mask], df[~mask], info_df[mask]\n",
    "\n",
    "\n",
    "def find_intergenic(df:pd.DataFrame, code_name:str='ecol', allowed_overlap:int=30, include_hypothetical_proteins:bool=False):\n",
    "    '''A GTDB ORF is intergenic if it either (1) does not overlap with any other nCDS element in the NCBI reference or (2) the overlap with a \n",
    "    protein (non-ppseudo) in the reference genome is no greater than the specified margin. \n",
    "    https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-15-721\n",
    "    https://pmc.ncbi.nlm.nih.gov/articles/PMC525685/\n",
    "    '''\n",
    "    gbff_df = GBFFFile(f'../data/model_organisms/ncbi_{code_name}_genomic.gbff').to_df(drop_pseudogenes=True) \n",
    "    if (not include_hypothetical_proteins):\n",
    "        gbff_df = gbff_df[gbff_df['product'].str.match('hypothetical protein')]\n",
    "\n",
    "    def is_intergenic(hit:pd.DataFrame, allowed_overlap:int=20) -> (bool, dict):\n",
    "        if hit is None:\n",
    "            return True, dict()\n",
    "        if hit['percent_overlap'] > 50:\n",
    "            return False, hit\n",
    "        if hit['overlap'] > allowed_overlap:\n",
    "            return False, hit\n",
    "        return True, dict()\n",
    "\n",
    "    info_df, mask = list(), list()\n",
    "    for row in tqdm(list(df.itertuples()), desc='find_intergenic: Looking for intergenic regions in the reference genome...'):\n",
    "        hit = search(gbff_df, start=int(row.start), stop=int(row.stop))\n",
    "        val, hit = is_intergenic(hit, allowed_overlap=allowed_overlap)\n",
    "        hit['id'] = row.Index\n",
    "        info_df.append(hit)\n",
    "        mask.append(val)\n",
    "    info_df = pd.DataFrame(info_df).set_index('id')\n",
    "    mask = np.array(mask)\n",
    "\n",
    "    print(f'find_intergenic: Found {np.sum(mask)} sequences in the input genome which are in intergenic regions.')\n",
    "    return df[mask], df[~mask], info_df[~mask]\n",
    "\n",
    "\n",
    "gtdb_fasta_df = FASTAFile('../data/model_organisms/gtdb_ecol_protein.faa').to_df(parser=fasta_file_parser_gtdb)[gtdb_fasta_cols]\n",
    "matches_df, gtdb_fasta_df, _ = find_matches(gtdb_fasta_df)\n",
    "boundary_errors_df, gtdb_fasta_df, boundary_errors_info_df = find_boundary_errors(gtdb_fasta_df)\n",
    "pseudogenes_df, gtdb_fasta_df, _ = find_pseudogenes(gtdb_fasta_df)\n",
    "intergenic_df, gtdb_fasta_df, _ = find_intergenic(gtdb_fasta_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>partial</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NZ_CP033092.2_2712</th>\n",
       "      <td>2859633</td>\n",
       "      <td>2859986</td>\n",
       "      <td>1</td>\n",
       "      <td>00</td>\n",
       "      <td>MPHTVHYPEQPSQGHVTPPPSRDAPVRFVLCCVAKTDNIRIYLVFH...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      start     stop strand partial  \\\n",
       "id                                                    \n",
       "NZ_CP033092.2_2712  2859633  2859986      1      00   \n",
       "\n",
       "                                                                  seq  \n",
       "id                                                                     \n",
       "NZ_CP033092.2_2712  MPHTVHYPEQPSQGHVTPPPSRDAPVRFVLCCVAKTDNIRIYLVFH...  "
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_fasta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>sequence_identity</th>\n",
       "      <th>alignment_length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>num_gap_openings</th>\n",
       "      <th>query_alignment_start</th>\n",
       "      <th>query_alignment_end</th>\n",
       "      <th>subject_alignment_start</th>\n",
       "      <th>subject_alignment_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>query_coverage_per_subject</th>\n",
       "      <th>query_coverage_per_pair</th>\n",
       "      <th>query_sequence_length</th>\n",
       "      <th>subject_sequence_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NZ_CP033092.2_12</th>\n",
       "      <td>NZ_CP033092.2_12</td>\n",
       "      <td>WP_000253469.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>430.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>430.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>445.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>862.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>445.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033092.2_15</th>\n",
       "      <td>NZ_CP033092.2_15</td>\n",
       "      <td>WP_001332265.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>86.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>7.400000e-62</td>\n",
       "      <td>180.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033092.2_36</th>\n",
       "      <td>NZ_CP033092.2_36</td>\n",
       "      <td>WP_011478308.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>396.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>412.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>396.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>776.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>413.0</td>\n",
       "      <td>396.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033092.2_56</th>\n",
       "      <td>NZ_CP033092.2_56</td>\n",
       "      <td>WP_001070179.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>218.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>218.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>229.0</td>\n",
       "      <td>1.560000e-165</td>\n",
       "      <td>452.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>229.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033092.2_145</th>\n",
       "      <td>NZ_CP033092.2_145</td>\n",
       "      <td>WP_001135732.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.600000e-32</td>\n",
       "      <td>103.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033091.2_97</th>\n",
       "      <td>NZ_CP033091.2_97</td>\n",
       "      <td>WP_042007484.1</td>\n",
       "      <td>99.187</td>\n",
       "      <td>123.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>3.180000e-87</td>\n",
       "      <td>247.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>153.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033091.2_105</th>\n",
       "      <td>NZ_CP033091.2_105</td>\n",
       "      <td>WP_094096600.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>5.940000e-92</td>\n",
       "      <td>262.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>232.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033091.2_109</th>\n",
       "      <td>NZ_CP033091.2_109</td>\n",
       "      <td>WP_012300440.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>141.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>4.510000e-101</td>\n",
       "      <td>283.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033091.2_112</th>\n",
       "      <td>NZ_CP033091.2_112</td>\n",
       "      <td>WP_094096600.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>5.940000e-92</td>\n",
       "      <td>262.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>232.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_CP033091.2_128</th>\n",
       "      <td>NZ_CP033091.2_128</td>\n",
       "      <td>WP_094096600.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>5.940000e-92</td>\n",
       "      <td>262.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>232.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            query_id      subject_id  sequence_identity  \\\n",
       "id                                                                        \n",
       "NZ_CP033092.2_12    NZ_CP033092.2_12  WP_000253469.1            100.000   \n",
       "NZ_CP033092.2_15    NZ_CP033092.2_15  WP_001332265.1            100.000   \n",
       "NZ_CP033092.2_36    NZ_CP033092.2_36  WP_011478308.1            100.000   \n",
       "NZ_CP033092.2_56    NZ_CP033092.2_56  WP_001070179.1            100.000   \n",
       "NZ_CP033092.2_145  NZ_CP033092.2_145  WP_001135732.1            100.000   \n",
       "...                              ...             ...                ...   \n",
       "NZ_CP033091.2_97    NZ_CP033091.2_97  WP_042007484.1             99.187   \n",
       "NZ_CP033091.2_105  NZ_CP033091.2_105  WP_094096600.1            100.000   \n",
       "NZ_CP033091.2_109  NZ_CP033091.2_109  WP_012300440.1            100.000   \n",
       "NZ_CP033091.2_112  NZ_CP033091.2_112  WP_094096600.1            100.000   \n",
       "NZ_CP033091.2_128  NZ_CP033091.2_128  WP_094096600.1            100.000   \n",
       "\n",
       "                   alignment_length  mismatch  num_gap_openings  \\\n",
       "id                                                                \n",
       "NZ_CP033092.2_12              430.0       0.0               0.0   \n",
       "NZ_CP033092.2_15               86.0       0.0               0.0   \n",
       "NZ_CP033092.2_36              396.0       0.0               0.0   \n",
       "NZ_CP033092.2_56              218.0       0.0               0.0   \n",
       "NZ_CP033092.2_145              50.0       0.0               0.0   \n",
       "...                             ...       ...               ...   \n",
       "NZ_CP033091.2_97              123.0       1.0               0.0   \n",
       "NZ_CP033091.2_105             125.0       0.0               0.0   \n",
       "NZ_CP033091.2_109             141.0       0.0               0.0   \n",
       "NZ_CP033091.2_112             125.0       0.0               0.0   \n",
       "NZ_CP033091.2_128             125.0       0.0               0.0   \n",
       "\n",
       "                   query_alignment_start  query_alignment_end  \\\n",
       "id                                                              \n",
       "NZ_CP033092.2_12                     1.0                430.0   \n",
       "NZ_CP033092.2_15                     1.0                 86.0   \n",
       "NZ_CP033092.2_36                    17.0                412.0   \n",
       "NZ_CP033092.2_56                     1.0                218.0   \n",
       "NZ_CP033092.2_145                   21.0                 70.0   \n",
       "...                                  ...                  ...   \n",
       "NZ_CP033091.2_97                     1.0                123.0   \n",
       "NZ_CP033091.2_105                    1.0                125.0   \n",
       "NZ_CP033091.2_109                    9.0                149.0   \n",
       "NZ_CP033091.2_112                    1.0                125.0   \n",
       "NZ_CP033091.2_128                    1.0                125.0   \n",
       "\n",
       "                   subject_alignment_start  subject_alignment_end  \\\n",
       "id                                                                  \n",
       "NZ_CP033092.2_12                      16.0                  445.0   \n",
       "NZ_CP033092.2_15                      25.0                  110.0   \n",
       "NZ_CP033092.2_36                       1.0                  396.0   \n",
       "NZ_CP033092.2_56                      12.0                  229.0   \n",
       "NZ_CP033092.2_145                      1.0                   50.0   \n",
       "...                                    ...                    ...   \n",
       "NZ_CP033091.2_97                      31.0                  153.0   \n",
       "NZ_CP033091.2_105                    108.0                  232.0   \n",
       "NZ_CP033091.2_109                      1.0                  141.0   \n",
       "NZ_CP033091.2_112                    108.0                  232.0   \n",
       "NZ_CP033091.2_128                    108.0                  232.0   \n",
       "\n",
       "                         e_value  bit_score  query_coverage_per_subject  \\\n",
       "id                                                                        \n",
       "NZ_CP033092.2_12    0.000000e+00      862.0                        99.0   \n",
       "NZ_CP033092.2_15    7.400000e-62      180.0                        99.0   \n",
       "NZ_CP033092.2_36    0.000000e+00      776.0                        96.0   \n",
       "NZ_CP033092.2_56   1.560000e-165      452.0                        99.0   \n",
       "NZ_CP033092.2_145   1.600000e-32      103.0                        70.0   \n",
       "...                          ...        ...                         ...   \n",
       "NZ_CP033091.2_97    3.180000e-87      247.0                        99.0   \n",
       "NZ_CP033091.2_105   5.940000e-92      262.0                        99.0   \n",
       "NZ_CP033091.2_109  4.510000e-101      283.0                        94.0   \n",
       "NZ_CP033091.2_112   5.940000e-92      262.0                        99.0   \n",
       "NZ_CP033091.2_128   5.940000e-92      262.0                        99.0   \n",
       "\n",
       "                   query_coverage_per_pair  query_sequence_length  \\\n",
       "id                                                                  \n",
       "NZ_CP033092.2_12                      99.0                  431.0   \n",
       "NZ_CP033092.2_15                      99.0                   87.0   \n",
       "NZ_CP033092.2_36                      96.0                  413.0   \n",
       "NZ_CP033092.2_56                      99.0                  219.0   \n",
       "NZ_CP033092.2_145                     70.0                   71.0   \n",
       "...                                    ...                    ...   \n",
       "NZ_CP033091.2_97                      99.0                  124.0   \n",
       "NZ_CP033091.2_105                     99.0                  126.0   \n",
       "NZ_CP033091.2_109                     94.0                  150.0   \n",
       "NZ_CP033091.2_112                     99.0                  126.0   \n",
       "NZ_CP033091.2_128                     99.0                  126.0   \n",
       "\n",
       "                   subject_sequence_length  \n",
       "id                                          \n",
       "NZ_CP033092.2_12                     445.0  \n",
       "NZ_CP033092.2_15                     110.0  \n",
       "NZ_CP033092.2_36                     396.0  \n",
       "NZ_CP033092.2_56                     229.0  \n",
       "NZ_CP033092.2_145                     50.0  \n",
       "...                                    ...  \n",
       "NZ_CP033091.2_97                     153.0  \n",
       "NZ_CP033091.2_105                    232.0  \n",
       "NZ_CP033091.2_109                    141.0  \n",
       "NZ_CP033091.2_112                    232.0  \n",
       "NZ_CP033091.2_128                    232.0  \n",
       "\n",
       "[180 rows x 16 columns]"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "boundary_errors_info_df[(boundary_errors_info_df.query_alignment_start == 1) & (boundary_errors_info_df.subject_alignment_start == 1)]\n",
    "boundary_errors_info_df[(boundary_errors_info_df.query_alignment_end == boundary_errors_info_df.query_sequence_length - 1) & (boundary_errors_info_df.subject_alignment_end == boundary_errors_info_df.subject_sequence_length)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GBFFFile.__init__: Found 4818 coding sequences in the GBFF file.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GBFFFile.__init__: Parsing GBFF file entries.: 100%|██████████| 4818/4818 [00:00<00:00, 11240.32it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>partial</th>\n",
       "      <th>product</th>\n",
       "      <th>frameshifted</th>\n",
       "      <th>incomplete</th>\n",
       "      <th>internal_stop</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>overlap</th>\n",
       "      <th>percent_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>130023</td>\n",
       "      <td>131519</td>\n",
       "      <td>00</td>\n",
       "      <td>L-xylulokinase</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>WP_000196108.1</td>\n",
       "      <td>MTQYWLGLDCGGSWLKAGLYDREGREAGVQRLPLCALSPQPGWAER...</td>\n",
       "      <td>296</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   strand   start    stop partial         product  frameshifted  incomplete  \\\n",
       "0      -1  130023  131519      00  L-xylulokinase         False       False   \n",
       "\n",
       "   internal_stop      protein_id  \\\n",
       "0          False  WP_000196108.1   \n",
       "\n",
       "                                                 seq  overlap  percent_overlap  \n",
       "0  MTQYWLGLDCGGSWLKAGLYDREGREAGVQRLPLCALSPQPGWAER...      296            100.0  "
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gbff_df = GBFFFile('../data/model_organisms/ncbi_ecol_genomic.gbff').to_df(drop_pseudogenes=True) # Load in the GBFF file again, but keep the pseudogenes and duplicate proteins. \n",
    "\n",
    "pd.DataFrame(search(gbff_df, start=130629, stop=130925, top_hit=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "selenobot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
