{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation\n",
    "\n",
    "In order to further validate the promising results of the PLM classifier on the testing data, we sought to interrogate a well-characterized microbial genome using the model. To this end, we chose to analyze the complete genome of Escherichia coli (strain K-12, substrain. MG1655, assembly ASM584v2). The RefSeq accession for this genome is GCF_000005845.2. \n",
    "\n",
    "For organizational purposes, all data used for model validation is placed into a subdirectory within the data directory. If you want to run this code, be sure to modify the `DATA_DIR` variable below to specify where the data will be stored on your machine.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "# Add the selenobot/ subirectory to the module search path, so that the modules in this directory are visible from the notebook.\n",
    "sys.path.append('../selenobot/')\n",
    "\n",
    "from dataset import Dataset\n",
    "from classifiers import Classifier\n",
    "from utils import csv_size, dataframe_from_gff, dataframe_to_fasta, dataframe_from_m8, dataframe_from_fasta\n",
    "from extend import extend\n",
    "import pandas as pd\n",
    "import re\n",
    "from typing import NoReturn, List\n",
    "import numpy as np\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = '/home/prichter/Documents/data/selenobot/validation/'\n",
    "\n",
    "# E. coli strain K-12 substrain MG1655\n",
    "genome_id = 'GCF_000005845.2' # The accession of the genome to download. \n",
    "assembly = 'ASM584v2' # The specific genome assembly.\n",
    "\n",
    "# https://www.nature.com/articles/s41396-022-01222-x#Sec2\n",
    "\n",
    "# Synergistota\n",
    "# genome_id = 'GCA_018398915.1' \n",
    "# assembly = 'ASM1839891v1'\n",
    "\n",
    "# Methanosarcinales archaeon \n",
    "# genome_id = 'GCA_026016205.1'\n",
    "# assembly = 'ASM2601620v1'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading genome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  561k    0  561k    0     0   157k      0 --:--:--  0:00:03 --:--:--  157k\n",
      "Archive:  /home/prichter/Documents/data/selenobot/validation/ncbi_dataset.zip\n",
      "  inflating: /home/prichter/Documents/data/selenobot/validation/README.md  \n",
      "  inflating: /home/prichter/Documents/data/selenobot/validation/ncbi_dataset/data/assembly_data_report.jsonl  \n",
      "  inflating: /home/prichter/Documents/data/selenobot/validation/ncbi_dataset/data/GCA_026016205.1/GCA_026016205.1_ASM2601620v1_genomic.fna  \n",
      "  inflating: /home/prichter/Documents/data/selenobot/validation/ncbi_dataset/data/GCA_026016205.1/protein.faa  \n",
      "  inflating: /home/prichter/Documents/data/selenobot/validation/ncbi_dataset/data/GCA_026016205.1/genomic.gff  \n",
      "  inflating: /home/prichter/Documents/data/selenobot/validation/ncbi_dataset/data/dataset_catalog.json  \n"
     ]
    }
   ],
   "source": [
    "# Download and unzip the genome data from NCBI. \n",
    "! curl 'https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/accession/{genome_id}/download?include_annotation_type=GENOME_FASTA,PROT_FASTA,GENOME_GFF' -o '{DATA_DIR}ncbi_dataset.zip'\n",
    "! unzip '{DATA_DIR}ncbi_dataset.zip' -d '{DATA_DIR}'\n",
    "\n",
    "# Create a directory to store the genome files. \n",
    "! mkdir '{DATA_DIR}{genome_id}/' \n",
    "# Move the relevant files into the new directory for organizational purposes. \n",
    "! mv '{DATA_DIR}ncbi_dataset/data/{genome_id}/genomic.gff' -t '{DATA_DIR}{genome_id}/'\n",
    "! mv '{DATA_DIR}ncbi_dataset/data/{genome_id}/protein.faa' -t '{DATA_DIR}{genome_id}/'\n",
    "! mv '{DATA_DIR}ncbi_dataset/data/{genome_id}/{genome_id}_{assembly}_genomic.fna' '{DATA_DIR}{genome_id}/genome.fna'\n",
    "\n",
    "# Remove some extraneous files which were also downloaded. \n",
    "! rm '{DATA_DIR}README.md'\n",
    "! rm -R '{DATA_DIR}ncbi_dataset'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean(path:str) -> NoReturn:\n",
    "    '''Modify the format of downloaded FASTA files to standardize reading and writing FASTA files.'''\n",
    "    fasta = ''\n",
    "    with open(path, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "    # In the FASTA files downloaded from NCBI, the only relevant information is right after the > character.\n",
    "    # This is the GenBank protein accession, which will be called the 'id'.\n",
    "    for line in lines:\n",
    "        if '>' in line: # This symbol marks the beginning of a header file.\n",
    "            id_ = re.search('>([^\\s]+)', line).group(1)\n",
    "            fasta += f'>id={id_}\\n'\n",
    "        else:\n",
    "            fasta += line\n",
    "    # This will overwrite the original file. \n",
    "    with open(path, 'w') as f:\n",
    "        f.write(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean(f'{DATA_DIR}{genome_id}/protein.faa')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Truncating known selenoproteins\n",
    "\n",
    "The E. coli genome assembly used contains three known selenoproteins: fdoG (RefSeq accession NP_418330.1), fdnG (RefSeq accession NP_415991.1) and fdhF (RefSeq accession NP_418503.1). The RefSeq accessions are found using the NCBI annotation browser for the particular E. coli genome assembly. Prior to generating the PLM embeddings, each of these sequences in the protein.faa file is truncated at the first selenocysteine residue. These sequences will act as sanity checks, ensuring that the PLM classifier is able to identify sequences known to be truncated selenoproteins. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_selenoproteins = ['NP_418330.1', 'NP_415991.1', 'NP_418503.1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def truncate_selenoproteins(path:str, gene_ids:List[str]=None) -> NoReturn:\n",
    "    '''Truncate the known selenoproteins stored in the input file, which are specified using the gene_ids\n",
    "    parameter.'''\n",
    "    # Load the selenoproteins into a pandas DataFrame. \n",
    "    df = dataframe_from_fasta(path)\n",
    "    for row in df.itertuples(index=True):\n",
    "        if getattr(row, 'id') in gene_ids:\n",
    "            u_loc = row.seq.index('U')\n",
    "            # setattr(row, 'seq', row.seq[:u_loc])\n",
    "            row = row._replace(seq=row.seq[:u_loc])\n",
    "            print(f\"truncate_selenoproteins: Truncated {getattr(row, 'id')}.\")\n",
    "            df.loc[row.Index] = row[1:] # Save the edited row in the DataFrame.\n",
    "    dataframe_to_fasta(df, path=path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "truncate_selenoproteins: Truncated NP_415991.1.\n",
      "truncate_selenoproteins: Truncated NP_418330.1.\n",
      "truncate_selenoproteins: Truncated NP_418503.1.\n"
     ]
    }
   ],
   "source": [
    "truncate_selenoproteins(f'{DATA_DIR}{genome_id}/protein.faa', gene_ids=known_selenoproteins)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Predicting selenoproteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating PLM embeddings\n",
    "\n",
    "Using the same procedure as was used to create embeddings for the UniProt sequences for training the classifier, embeddings are generated for all predicted genes in the model genome. The code for embedding sequence data is provided in the `embed.py` file in the `scripts` directory. The process for generating embeddings was computationally-intensive, and had to be run on an external computer cluster. The steps of the embedding algorithm, which is implemented by the `PlmEmbedder` class, are given below.\n",
    "\n",
    "1. **Amino acid sequences are read from a FASTA file.**\n",
    "2. **All non-standard amino acids were replaced with an “X”.**\n",
    "3. **Sequences were sorted in ascending order according to length.** This avoids the addition of unnecessary padding. \n",
    "4. **The sequences are processed and tokenized in batches.** Processing sequences in batches enables one to generate the embeddings as quickly as possible, while also preventing the GPUs from crashing. The maximum number of sequences in any given batch is 100, the maximum number of amino acids in a batch is 4000, and any sequence longer than 1000 amino acids is processed individually. \n",
    "5. **The PLM is used to generate embeddings.** These embeddings have shape shape `(length, latent dimension)`.\n",
    "6. **Each embedding is sliced according to the length of the original sequence.** This is due to the fact that part of the model output corresponds to the padding tokens, which should be excluded from the final embedding. \n",
    "7. **The embeddings are mean-pooled over sequence length.** This step standardizes the length of the embedding vectors to be of fixed dimension (the latent dimension of the PLM), which is necessary for passing them to the Selenobot linear classifier. \n",
    "The output of the steps above is data containing gene IDs, the original amino acid sequences (which may contain non-standard residues), and columns with the mean-pooled embeddings. \n",
    "\n",
    "There are several pre-embedded genomes available in the Google Cloud repository, which can be downloaded using the code below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "curl 'https://storage.googleapis.com/selenobot-data/{genome_id}.csv' -o '{DATA_DIR}{genome_id}.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running the PLM classifier\n",
    "\n",
    "The same workflow as described in the `training.ipynb` and `testing.ipynb` notebooks is used to predict selenoproteins in the model genome. First, the `proteins.faa` file, which contains the predicted amino acid sequences in the genome, is loaded into a pandas `DataFrame` and used to initialize a `Dataset` object. Then, the model’s predict method is called on the `Dataset`, generating a `TestReporter` object which stores model output. A threshold of `0.5` is applied to the model output, selecting positive classifications in which the model is over 50 percent confident. The IDs of the candidate selenoprotein sequences and the model confidence values are written to a `predictions.csv` file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(f'{DATA_DIR}{genome_id}.csv', index_col=0)\n",
    "# df = pd.read_csv(f'{DATA_DIR}../test.csv', index_col=0)\n",
    "dataset = Dataset(df, embedder=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<All keys matched successfully>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = Classifier(latent_dim=1024, hidden_dim=512)\n",
    "model.load_state_dict(torch.load(f'{DATA_DIR}../plm_model_weights.pth'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of predicted selenoproteins: 3\n",
      "Model confidence for each predicted selenoprotein: [0.9749483  0.8699334  0.57687116]\n"
     ]
    }
   ],
   "source": [
    "reporter = model.predict(dataset)\n",
    "\n",
    "# Apply the threshold to model outputs, and find the sequence IDs in the proteome for which the prediciton is 1.\n",
    "predictions = df[reporter.apply_threshold(threshold=0.5) == 1]\n",
    "# Add the model confidences to the DataFrame. \n",
    "predictions['confidence'] = reporter.outputs[reporter.apply_threshold(threshold=0.5) == 1]\n",
    "\n",
    "# Print some information about the hits: how many hits, and how confident is the model that they are truncated selenoproteins?\n",
    "print('Number of predicted selenoproteins:', len(predictions))\n",
    "print('Model confidence for each predicted selenoprotein:', predictions.confidence.values)\n",
    "\n",
    "# Write the predicted selenoproteins to a file, along with the model confidence. \n",
    "predictions[['id', 'confidence']].set_index('id').to_csv(f'{DATA_DIR}{genome_id}/predictions.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extending flagged sequences\n",
    "\n",
    "First, the data in the genomic.gff file (e.g.  the start and stop position of each sequence in the genome, as well as whether the gene is read from the forward or reverse strand) is merged with the DataFrame containing the predicted selenoproteins. This information is used to extend each amino acid sequence to the next stop codon using the following procedure.\n",
    "\n",
    "1. **Read in the complete genome sequence from the genome.fna file.**\n",
    "2. **The terminal location of the sequence is located in the genome, and incremented from right to left (reverse) or left to right (forward) until the next stop codon is encountered.** Valid stop codons are TAA, TAG and TGA, and their reversed equivalents.\n",
    "3. **The new gene coordinates, the original stop codon, etc. are stored in the input DataFrame.** \n",
    "5. **Each extended nucleotide sequence is translated to a sequence of amino acids.** This is done using the `BioPython` `Seq` module. The translation utilities provided by `BioPython` do not support non-standard amino acids, so the selenocysteine residue (U) is added manually after the translation step. \n",
    "\n",
    "The extended sequences are written to a FASTA file called `query.faa`, which is used during homology search with MMSeqs2. In addition to the extended sequences, this file contains the following information in the header fields of each entry. \n",
    "1. **gene_id** The RefSeq gene ID of the query sequence (e.g. fdoG)\n",
    "2. **id** The GenBank unique identifier of the coding sequence. \n",
    "3. **nt_ext** How much the sequence was extended, in terms of the number of nucleotides. \n",
    "4. **aa_ext** How much the sequence was extended, in terms of the number of amino acids. \n",
    "5. **nt_start** The start location of the gene encoding the protein. \n",
    "6. **nt_stop** The stop location of the gene encoding the protein. \n",
    "7. **reverse** Whether or not the gene is read from the forward or reverse strand. Either True or False. \n",
    "8. **u_codon** The codon encoding the putative selenocysteine residue (i.e. the original stop codon). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicted selenoprotein IDs:\n",
      "UZE92041.1 (None)\n",
      "UZE92443.1 (None)\n",
      "UZE92782.1 (None)\n"
     ]
    }
   ],
   "source": [
    "predictions = pd.read_csv(f'{DATA_DIR}{genome_id}/predictions.csv') # Read in the model predictions. \n",
    "# Add the information from the GFF file to the DataFrame of predicted selenoproteins. \n",
    "gff_df = dataframe_from_gff(f'{DATA_DIR}{genome_id}/genomic.gff') \n",
    "predictions = predictions.merge(gff_df, on='id')\n",
    "\n",
    "print('Predicted selenoprotein IDs:\\n' + '\\n'.join([f\"{getattr(row, 'id')} ({row.gene_id})\" for row in predictions.itertuples()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_genome(path:str) -> str:\n",
    "    '''Load in the complete nucleotide sequence of the genome.\n",
    "    \n",
    "    :param path: A FASTA file from NCBI which contains a complete genome. \n",
    "    :return: A string of nucleotides. \n",
    "    '''\n",
    "    with open(path, 'r') as f:\n",
    "        lines = f.read().splitlines()[1:] # Skip the header line. \n",
    "        seq = ''.join(lines)\n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the complete genome of the model organism. \n",
    "genome = load_genome(f'{DATA_DIR}{genome_id}/genome.fna')\n",
    "\n",
    "# QVL37563.1, QVL37671.1 do not have a valid start codon, so removing. \n",
    "predictions = predictions[~predictions['id'].str.match('QVL37563.1')]\n",
    "predictions = predictions[~predictions['id'].str.match('QVL37671.1')]\n",
    "\n",
    "# Extend the predictions, and write them to a query.faa file for use with MMSeqs.\n",
    "predictions = extend(predictions, genome)\n",
    "dataframe_to_fasta(predictions, f'{DATA_DIR}{genome_id}/query.faa')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Homology search against GTDB\n",
    "\n",
    "The MMSeqs2 tool is used to search for sequences homologous to the extended candidate selenoproteins in the `query.faa` file. This search was performed on a remote computing cluster, and a full description of the procedure is provided in the Methodology document. The output of the homology search is a file where each row represents a single query-target match. The table has twelve columns:\n",
    "\n",
    "1. **The header from the `query.fasta` file.** This contains information like the query gene ID, sequence length, etc.\n",
    "2. **GTDB identifiers for the target sequence.**\n",
    "3. **Sequence identity.**\n",
    "4. **Alignment length.**\n",
    "5. **Number of mismatches.**\n",
    "6. **Number of gap openings.**\n",
    "7. **The start position of the matching domain in the query sequence.** This is given in terms of the number of amino acids from the start of the query sequence.\n",
    "8. **The stop position of the matching domain in the query sequence.** This is given in terms of the number of amino acids from the start of the query sequence.\n",
    "9. **The start position of the matching domain in the target sequence.** This is given in terms of the number of amino acids from the start of the query sequence.\n",
    "10. **The stop position of the matching domain in the target sequence.** This is given in terms of the number of amino acids from the start of the query sequence.\n",
    "11. **E-value.** This is the “expect value,” and reflects the likelihood of the sequence match occurring by chance.\n",
    "12. **Bit score.** This score is calculated from a formula that takes into account the alignment of similar or identical residues, as well as any gaps introduced to align the sequences. \n",
    "\n",
    "### Downloading sequences for MMSeqs2 hits\n",
    "\n",
    "The output file produced by MMSeqs2 does not contain the amino acid sequences for the matching targets. These sequences are stored in a remote SQL database hosted by Caltech, and were downloaded using the [Find-A-Bug Python API](https://github.com/pipparichter/find-a-bug-api). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install the Find-A-Bug API.\n",
    "# ! pip install -e /home/prichter/Documents/find-a-bug-api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import fabapi # Import the Find-A-Bug Python API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of MMSeqs hits for GCA_026016205.1 in the GTDB database: 15357\n"
     ]
    }
   ],
   "source": [
    "# Load in the MMSeqs results, dropping the two columns which are not populated by the program.\n",
    "mmseqs_df = dataframe_from_m8(f'{DATA_DIR}{genome_id}/results.m8').drop(columns=['query_align_seq', 'target_align_seq'])\n",
    "print(f'Number of MMSeqs hits for {genome_id} in the GTDB database:', len(mmseqs_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the target sequences using the Find-A-Bug API.\n",
    "target_seqs = []\n",
    "for target_id in mmseqs_df.target_id:\n",
    "    # get_sequence_with_gene_id returns a DataFrame, so we need to extract the sequence.\n",
    "    seq = fabapi.get_sequence_with_gene_id(target_id).seq.item()\n",
    "    target_seqs.append(seq)\n",
    "mmseqs_df['target_seq'] = target_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the new DataFrame (with sequences added) to a file to avoid re-generating in case the kernel crashes.\n",
    "mmseqs_df.set_index('query_id').to_csv(f'{DATA_DIR}{genome_id}/results.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering MMSeqs2 hits\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "mmseqs_df = pd.read_csv(f'{DATA_DIR}{genome_id}/results.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_hit_overlap(row):\n",
    "    '''Check if a particular MMSeqs hit spans the stop codon in the query sequence.'''\n",
    "    # We want the start of the hit to precede the U codon, and the end of the hit\n",
    "    # to lie in the extended region. \n",
    "    u_pos = int(row.query_aa_length) - (int(row.query_nt_ext) // 3)\n",
    "    return (u_pos > row.target_align_start) and (u_pos < row.target_align_stop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of hits before filtering by overlap: 15357\n",
      "Number of hits after filtering by overlap: 80\n"
     ]
    }
   ],
   "source": [
    "# We only care about the hits which overlap on the extended region.\n",
    "print('Number of hits before filtering by overlap:', len(mmseqs_df))\n",
    "mmseqs_df = mmseqs_df[mmseqs_df.apply(check_hit_overlap, axis=1)]\n",
    "print('Number of hits after filtering by overlap:', len(mmseqs_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean length of hits for UZE92782.1: 385.18666666666667\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAHFCAYAAADcytJ5AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAAA8Z0lEQVR4nO3deVgVdf//8dcB4bAJuYGgKC6V5J5auaVmt90uZZaaZaWZtrhl+i1Ty6VU2m/LOy2t1Eqzus0270orpUxNRc2lbnJDzSV3cWP//P7gx9EjiBziMCM8H9c11zVn5jMz75lBeDmfmTkOY4wRAACADflYXQAAAMDFEFQAAIBtEVQAAIBtEVQAAIBtEVQAAIBtEVQAAIBtEVQAAIBtEVQAAIBtEVQAAIBtEVTgZvbs2XI4HFq7dq3VpeRp3759Gj9+vDZs2JBrXt++fRUSElLk24yJiVHfvn1dn5ctWyaHw6Fly5Z5tJ5p06Zp9uzZHi2T17a8sZ8rVqzQ+PHjdfz48Vzz2rZtq7Zt2xbp9gpq/fr1atOmjcLCwuRwODRlyhSvbSspKUkOh0Mvv/xynvNffvllORwOJSUlSco+Dw6HI98hJibGtfyl2p9v//796tu3r8LDwxUQEKAGDRronXfeyVXTp59+qrvvvlu1a9dWYGCgYmJi1Lt3b23dujXPfbvY8M9//tOt/bZt23TfffepWrVqCgwMVK1atTR8+HAdOXLErd3bb7+t22+/XTExMQoMDFTt2rX16KOPav/+/QU65suXL1f//v3VpEkTOZ1Ot+NbEA6HQ4MHD75ku5zfa+eve968eV79eULRKWN1AYAn9u3bpwkTJigmJkaNGjWypIZrr71WK1eu1DXXXOPRctOmTVPFihXdQo+3tuWpFStWaMKECerbt6+uuOIKt3nTpk3z6rbz069fP50+fVrz589XuXLl3P7wW+2ZZ57RI488kue82bNn66233lK3bt3cpgcGBuqHH37Id70nTpxQq1atlJaWphdffFGRkZH68MMP1b9/f504cULDhw93tX3hhRdUuXJljRkzRjVr1tSePXs0efJkXXvttVq1apXq1q0rSYqMjNTKlStzbeuzzz7TCy+84FbnoUOHdMMNNyg0NFTPPfecqlWrpvXr12vcuHFaunSpEhIS5OOT/X/ccePGqV27dpo8ebKqVKmixMREPffcc/r888+1fv16RURE5Luv33//vb777js1btxYoaGhHof/gurcubNWrlypyMhI17R58+Zp8+bNGjZsmFe2iaJDUAE8FBoaqhtuuMGr20hPT5fD4SiWbV2Kt0NSfjZv3qwBAwaoY8eORbK+nONapszf/9VXq1Yt1apVK9f0VatWadasWbrxxhv10ksvuc3z8fG55PmcPn26duzYobVr16pJkyaSpFtuuUX79+/X2LFj1a9fP1eY/PLLLxUeHu62/E033aSYmBj961//0ttvvy1JcjqdeW531KhRCgoK0t133+2a9vnnn+vIkSP66KOP1L59e0lSu3btlJqaqtGjR+vXX39V48aNJWVf8Tp/+23atNG1116rZs2aaebMmXr66afz3ddnnnlG48aNk5R9xcpbQaVSpUqqVKmSV9YN76PrB4WydetW3XPPPQoPD5fT6VRsbKzeeOMNtzY53RYffvihxowZo6ioKIWGhurmm29WYmKiW1tjjCZPnqzq1asrICBATZs21ZIlS9y6HZYtW6ZmzZpJkh544AHXZevx48e7rWvbtm3q1KmTQkJCFB0drREjRig1NfWS+5Senq4nn3xSlStXVlBQkFq1aqXVq1fnapdXd8yOHTvUq1cvRUVFyel0KiIiQu3bt3d1UcXExGjLli2Kj4/P1S2Qs773339fI0aMUJUqVeR0OrVt27Z8u5m2bNmi9u3bKzg4WJUqVdLgwYN15swZ1/ycy/15dTedf9zGjx+vJ554QpJUo0YNV30528yr6+fo0aMaOHCgqlSpIn9/f9WsWVNjxozJdZxzLs2///77io2NVVBQkBo2bKivvvrq4idC5y7VZ2RkaPr06bm6RzZv3qyuXbuqXLlyCggIUKNGjTRnzhy3deR3XL3lwIEDuvPOO1WpUiV9/PHHhQpEP//8syIiIlwhJUeXLl10+vRpffPNN65pF4YUSYqKilLVqlW1Z8+efLezfft2xcfHq2fPngoNDXVN9/PzkySFhYW5tc8JRwEBAfluv0mTJvL19b3k9iW5rsz8XZf6+bqw66dt27ZatGiRdu3alWf32/Tp09WwYUOFhISobNmyqlOnjkaPHl0ktcJzXFGBx3777Te1aNFC1apV0yuvvKLKlSvr22+/1dChQ3X48GHX/5ByjB49Wi1bttTbb7+t5ORkjRw5Urfeeqt+//13+fr6SpLGjBmjuLg4PfTQQ7rjjju0Z88e9e/fX+np6brqqqskZXeDzJo1Sw888ICefvppde7cWZJUtWpV17bS09N122236cEHH9SIESP0448/6rnnnlNYWJjGjh2b734NGDBA7733nv7v//5P//jHP7R582bdcccdOnny5CWPSadOnZSZmakXX3xR1apV0+HDh7VixQrXPR8LFy5U9+7dFRYW5upKcTqdbusYNWqUmjdvrjfffFM+Pj4KDw/XgQMH8txeenq6OnXqpIcfflhPPfWUVqxYoYkTJ2rXrl368ssvL1nv+fr376+jR49q6tSp+vTTT12Xxy92JSUlJUXt2rXT9u3bNWHCBDVo0EA//fST4uLitGHDBi1atMit/aJFi7RmzRo9++yzCgkJ0Ysvvqhu3bopMTFRNWvWzHMbOZfqmzdvru7du2vEiBGueYmJiWrRooXCw8P1+uuvq0KFCvrggw/Ut29f/fXXX3ryySfd1pXXcfWG9PR09ejRQ4cPH1Z8fPxFuz0yMjJyTfPx8XH90U5LS8v1syGd+3nZuHGjevXqddE6duzYoV27dun222/Pt953331Xxhj179/fbfrtt9+uatWqacSIEZo2bZqqV6+udevW6fnnn9ett96q2NjYfNcbHx+vzMxMV7eTtxXm52vatGl66KGHtH37di1cuNBt3vz58zVw4EANGTJEL7/8snx8fLRt2zb99ttvxbE7yIsBzjNr1iwjyaxZs+aibW655RZTtWpVc+LECbfpgwcPNgEBAebo0aPGGGOWLl1qJJlOnTq5tfv444+NJLNy5UpjjDFHjx41TqfT3HXXXW7tVq5caSSZNm3auKatWbPGSDKzZs3KVVefPn2MJPPxxx+7Te/UqZO5+uqr893v33//3Ugyjz/+uNv0uXPnGkmmT58+rmk5+7V06VJjjDGHDx82ksyUKVPy3UbdunXd9uXC9d14440XnZezrfP387XXXnNrO2nSJCPJLF++3BhjzM6dOy96rCSZcePGuT6/9NJLRpLZuXNnrrZt2rRxq/vNN9/M8zi/8MILRpJZvHix23YiIiJMcnKya9qBAweMj4+PiYuLy7WtvOocNGiQ27RevXoZp9Npdu/e7Ta9Y8eOJigoyBw/ftwYk/9xzUvO8XrppZfynJ/fMTLGmIEDBxpJ5s0338xzfs55y2to3769q92wYcOMj4+P2bVrl9vy9913n5FkHnrooYvuQ3p6umnbtq0JDQ3NdXzOl5GRYapUqWLq1KmT5/x9+/aZ5s2bu9XYo0cPk5KSctF1GmNMcnKyiY2NNdHR0ebkyZP5tr3QpY5vXgr685Xze+38dXfu3NlUr1491zoHDx5srrjiCo9qh3fR9QOPpKSk6Pvvv1e3bt0UFBSkjIwM19CpUyelpKRo1apVbsvcdtttbp8bNGggSdq1a5ek7D791NRU9ezZ063dDTfc4PHNkw6HQ7feemuu7eVs62KWLl0qSerdu7fb9J49e17y8n358uVVq1YtvfTSS3r11Ve1fv16ZWVleVS3JN15550etb+w1nvuuUfSuX3xlh9++EHBwcHq3r272/Scm4S///57t+nt2rVT2bJlXZ8jIiIUHh5+yXOS3/bbt2+v6OjoXNs/c+ZMrptGPT2uhTF79mxNmzZN/fr108MPP3zRdoGBgVqzZk2u4fwblh966CH5+fmpd+/e2rJli44cOaI33nhDH330kaSLd5cYY/Tggw/qp59+0nvvvZfr+Jzvm2++0d69e/Xggw/mmnfs2DF17dpVycnJmjt3rn788UdNmzZNy5cv12233ZbnFSEp+3fDHXfcoV27dumTTz7xyhN4eSnqn6/rrrtOx48f1913363PP/9chw8fLqpSUUh0/cAjR44cUUZGhqZOnaqpU6fm2ebCf9gVKlRw+5xzCfvs2bOudUrK81L5pZ4auFBQUJBbH3rO9lJSUvJdLqeGypUru00vU6ZMrvov5HA49P333+vZZ5/Viy++qBEjRqh8+fLq3bu3Jk2a5PZLND/nP5FwKXnVlVP7hY+QFrUjR46ocuXKuR6pDQ8PV5kyZXJtP6/j53Q6Xee/MNvP61hFRUW55p+voMc1J5BmZmbmOT/nD3TOPRw51q5dq0cffVRNmza95BNSPj4+atq0ab5tYmNjtXDhQj388MOqV6+eJCk6OlqvvPKKhgwZoipVquRaxvz/LpwPPvhAc+bMUdeuXfPdxjvvvCM/Pz/df//9uea98MIL2rBhg3bt2uU6dq1bt1adOnV00003ae7cuerTp4/bMqmpqerWrZuWL1+ur776Stdff32+2y9KRf3zdd999ykjI0MzZ87UnXfeqaysLDVr1kwTJ07UP/7xj79bLgqBKyrwSLly5eTr66u+ffvm+T/DNWvWqFOnTh6tM+cXzV9//ZVr3sXu0ShqOTVcuL2MjIwC/eGvXr263nnnHR04cECJiYl6/PHHNW3aNNdNqgVx4R/+/ORVV07tOfuSE9guvMH17waZChUq6K+//pIxxm36wYMHlZGRoYoVK/6t9Rdk+3m9p2Pfvn2SlGv7BT2uFStWlK+vr/bu3Zvn/L1798rX19ftD+OhQ4d0xx13KCQkRAsWLMjz3pLC6Nixo3bt2qU//vhDv/32m3bu3Ona7o033ujWNiekzJo1S2+//bbuvffefNd98OBBffXVV7rtttvyvF9nw4YNqlKlSq6Al3Mj++bNm92mp6am6vbbb9fSpUv12WefuZ4Uupw98MADWrFihU6cOKFFixbJGKMuXboU+ioN/h6CCjwSFBSkdu3aaf369WrQoIGaNm2aa7jUFYgLXX/99XI6na5L2zlWrVqV6xfDhVdjikrOUy1z5851m/7xxx9f9FL3xVx11VV6+umnVb9+fa1bt841/e/8Ly8vF9Y6b948Sef2JSIiQgEBAdq4caNbu88//zzXujw5ru3bt9epU6f02WefuU1/7733XPO9qX379vrhhx9cweT87QcFBRX6ce6AgAC1bNlSX3zxRa4rcCkpKfriiy/UqlUrVwDMyMhQjx49tG/fPn300UeqVq1a4XboIhwOh6688krFxsYqMzNTr732mho1auQWVIwxGjBggGbNmqW33npLDzzwwCXX+9577yk9PT3Pbh8p+8rUn3/+mSuw5XSpnX/zes6VlB9++EELFizQLbfcUphdtURB/j0GBwerY8eOGjNmjNLS0rRly5Ziqg7no+sHefrhhx/yfENkp06d9Nprr6lVq1Zq3bq1Hn30UcXExOjkyZPatm2bvvzyy0u+0OpC5cuX1/DhwxUXF6dy5cqpW7du+vPPPzVhwgRFRka69cnXqlVLgYGBmjt3rmJjYxUSEqKoqCjXZf/Cio2N1b333qspU6bIz89PN998szZv3qyXX37Z7dHNvGzcuFGDBw9Wjx49dOWVV8rf318//PCDNm7cqKeeesrVrn79+po/f74++ugj1axZUwEBAapfv36h6vX399crr7yiU6dOqVmzZq6nfjp27KhWrVpJyv5Dd++99+rdd99VrVq11LBhQ61evdoVaM6XU8drr72mPn36yM/PT1dffXWe3Vb333+/3njjDfXp00dJSUmqX7++li9frsmTJ6tTp066+eabC7VPBTVu3Dh99dVXateuncaOHavy5ctr7ty5WrRokV588cVcj9V64vnnn1e7du3UvHlzDRs2TNWqVdPu3bs1ZcoU/fXXX5o/f76r7RNPPKH4+Hj17t1bQUFBue7NynF+cMrKyrpou8aNG7sC45AhQ9S2bVtVqFBBO3bs0Ouvv64///xT8fHxbssMHTpU77zzjvr166f69eu7rdvpdLred3K+d955R9HR0RcNFYMGDdLcuXP1j3/8Q0899ZSio6O1efNmTZw4UREREW73RnXv3l1ff/21xowZowoVKrhtPzQ01O3Jsdq1a0uS2+Phhw4dcu3Tpk2bJElff/21670nbdq0ybPGolC/fn19+umnmj59upo0aeLqlhswYIACAwPVsmVLRUZG6sCBA4qLi1NYWJjrqhKKmZV38sJ+cu6Ov9iQc9f8zp07Tb9+/UyVKlWMn5+fqVSpkmnRooWZOHGia105T1188sknbtvI62mUrKwsM3HiRFO1alXj7+9vGjRoYL766ivTsGFD061bN7flP/zwQ1OnTh3j5+fn9vRKnz59THBwcK59GjdunCnIj3pqaqoZMWKECQ8PNwEBAeaGG24wK1euNNWrV8/3qZ+//vrL9O3b19SpU8cEBwebkJAQ06BBA/Ovf/3LZGRkuJZLSkoyHTp0MGXLljWSXE8cXOw45bWt8/dz48aNpm3btiYwMNCUL1/ePProo+bUqVNuy584ccL079/fREREmODgYHPrrbeapKSkXE/9GGPMqFGjTFRUlPHx8XHb5oVP/RhjzJEjR8wjjzxiIiMjTZkyZUz16tXNqFGjcj0Vojye2jHG5DqmF3Ox5Tdt2mRuvfVWExYWZvz9/U3Dhg1zPd2U33HNz9q1a023bt1MxYoVja+vr6lYsaLp1q2bSUhIyLUP+f1byRly5PfUjySzdetWV9uuXbuayMhI4+fnZypXrmz69u1rkpKSctWaXw15PdHy888/G0lm7Nix+R6DdevWmW7dupmqVasap9Npatasafr375/rSaL89ufCn5nq1avnqinnHBVk+bwU9Ocrr6d+jh49arp3726uuOIK43A4XOdqzpw5pl27diYiIsL4+/ubqKgo07NnT7Nx48ZL1gPvcBhzQUczYBM7d+5UnTp1NG7cOF62BAClFEEFtvDrr7/qww8/VIsWLRQaGqrExES9+OKLSk5O1ubNmz1++gcAUDJwjwpsITg4WGvXrtU777yj48ePKywsTG3bttWkSZMIKQBQinFFBQAA2BaPJwMAANsiqAAAANsiqAAAANu6rG+mzcrK0r59+1S2bFmPXj8OAACsY4zRyZMnFRUVddEv2sxxWQeVffv25fsNoQAAwL727Nnj9rUMebmsg0rO67337NlzydecF5nTp6Wc17Xv2ycFBxfPdgEAKCGSk5MVHR1doG+Xv6yDSk53T2hoaPEFFV/fc+OhoQQVAAAKqSC3bXAzLQAAsC2CCgAAsK3LuuvHEmXKSH36nBsHAABew19aTzmd0uzZVlcBAECpQNcPAACwLa6oeMoY6cyZ7PGgIIkXzQEA4DVcUfHUmTNSSEj2kBNYAACAVxBUAACAbRFUAACAbRFUAACAbRFUAACAbRFUAACAbRFUAACAbfEeFU/5+krdu58bBwAAXkNQ8VRAgPTJJ1ZXAQBAqUBQ+Ztinlrk9jnp+c4WVQIAQMnDPSoAAMC2CCqeOn06+/t9HI7scQAA4DUEFQAAYFsEFQAAYFsEFQAAYFsEFQAAYFsEFQAAYFsEFQAAYFu88M1Tvr5Sp07nxgEAgNcQVDwVECAtWnTpdgAA4G+j6wcAANgWQQUAANgWQcVTp09LwcHZA6/QBwDAq7hHpTDOnLG6AgAASgWuqAAAANsiqAAAANsiqAAAANsiqAAAANsiqAAAANviqR9P+fhIbdqcGwcAAF5DUPFUYKC0bJnVVQAAUCpwSQAAANgWQQUAANgWQcVTp09LlSplD7xCHwAAr+IelcI4fNjqCgAAKBW4ogIAAGyLoAIAAGyLoAIAAGyLoAIAAGyLoAIAAGyLp3485eMjNW16bhwAAHgNQcVTgYHSmjVWVwEAQKnAJQEAAGBbBBUAAGBblgaVjIwMPf3006pRo4YCAwNVs2ZNPfvss8rKyrKyrPydOSPFxGQPZ85YXQ0AACWapfeovPDCC3rzzTc1Z84c1a1bV2vXrtUDDzygsLAwPfbYY1aWdnHGSLt2nRsHAABeY2lQWblypbp27arOnTtLkmJiYvThhx9q7dq1VpYFAABswtKun1atWun777/XH3/8IUn69ddftXz5cnXq1CnP9qmpqUpOTnYbAABAyWXpFZWRI0fqxIkTqlOnjnx9fZWZmalJkybp7rvvzrN9XFycJkyYUMxVAgAAq1h6ReWjjz7SBx98oHnz5mndunWaM2eOXn75Zc2ZMyfP9qNGjdKJEydcw549e4q5YgAAUJwsvaLyxBNP6KmnnlKvXr0kSfXr19euXbsUFxenPn365GrvdDrldDqLu0wAAGARS4PKmTNn5HPBa+h9fX3t/XiywyFdc825cQAA4DWWBpVbb71VkyZNUrVq1VS3bl2tX79er776qvr162dlWfkLCpK2bLG6CgAASgVLg8rUqVP1zDPPaODAgTp48KCioqL08MMPa+zYsVaWBQAAbMLSoFK2bFlNmTJFU6ZMsbIMAABgU3zXj6fOnJHq1s0eeIU+AABeZekVlcuSMdJvv50bBwAAXsMVFQAAYFsEFQAAYFsEFQAAYFsEFQAAYFsEFQAAYFs89eMph0OqXv3cOAAA8BqCiqeCgqSkJKurAACgVKDrBwAA2BZBBQAA2BZBxVNnz0rNmmUPZ89aXQ0AACUa96h4KitLWrv23DgAAPAarqgAAADbIqgAAADbIqgAAADbIqgAAADbIqgAAADb4qmfwqhY0eoKAAAoFQgqngoOlg4dsroKAABKBbp+AACAbRFUAACAbRFUPHX2rNS2bfbAK/QBAPAq7lHxVFaWFB9/bhwAAHgNV1QAAIBtEVQAAIBtEVQAAIBtEVQAAIBtEVQAAIBt8dRPYQQFWV0BAAClAkHFU8HB0unTVlcBAECpQNcPAACwLa6oFLGYpxblmpb0fGcLKgEA4PLHFRVPpaRInTtnDykpVlcDAECJxhUVT2VmSv/977lxAADgNVxRAQAAtkVQAQAAtkVQAQAAtkVQAQAAtkVQAQAAtkVQAQAAtsXjyZ4KDpaMsboKAABKBa6oAAAA2yKoAAAA2yKoeColRerRI3vgFfoAAHgVQcVTmZnSf/6TPfAKfQAAvIqgAgAAbIugAgAAbIugAgAAbIugAgAAbIugAgAAbIugAgAAbItX6HsqKEg6dercOAAA8BqCiqccjuzv+wEAAF5H1w8AALAtgoqnUlOlvn2zh9RUq6sBAKBEI6h4KiNDmjMne8jIsLoaAABKNIIKAACwLYIKAACwLYIKAACwLYIKAACwLYIKAACwLYIKAACwLd5M66mgIOngwXPjAADAawgqnnI4pEqVrK4CAIBSga4fAABgWwQVT6WmSoMGZQ+8Qh8AAK+yPKjs3btX9957rypUqKCgoCA1atRICQkJVpd1cRkZ0rRp2QOv0AcAwKssvUfl2LFjatmypdq1a6evv/5a4eHh2r59u6644gorywIAADZhaVB54YUXFB0drVmzZrmmxcTEWFcQAACwFUu7fr744gs1bdpUPXr0UHh4uBo3bqyZM2detH1qaqqSk5PdBgAAUHJZekVlx44dmj59uoYPH67Ro0dr9erVGjp0qJxOp+6///5c7ePi4jRhwgQLKi1aMU8tyjUt6fnOFlQCAIC9WXpFJSsrS9dee60mT56sxo0b6+GHH9aAAQM0ffr0PNuPGjVKJ06ccA179uwp5ooBAEBxsjSoREZG6pprrnGbFhsbq927d+fZ3ul0KjQ01G0AAAAll6VdPy1btlRiYqLbtD/++EPVq1e3qKICCAyUdu48Nw4AALzG0qDy+OOPq0WLFpo8ebJ69uyp1atXa8aMGZoxY4aVZeXPx0fiySQAAIqFpV0/zZo108KFC/Xhhx+qXr16eu655zRlyhT17t3byrIAAIBNWP6lhF26dFGXLl2sLqPg0tKkMWOyxydNsrYWAABKOMtfoX/ZSU+XXn45e0hPt7oaAABKNIIKAACwLYIKAACwLYIKAACwLYIKAACwLYIKAACwLYIKAACwLcvfo3LZCQyUNm8+Nw4AALyGoOIpHx+pbl2rqwAAoFSg6wcAANgWV1Q8lZYmTZ6cPT56tLW1AABQwhFUPJWeLk2YkD3+xBPW1gIAQAlH1w8AALAtggoAALAtggoAALAtggoAALAtggoAALAtggoAALAtHk/2VECAtHr1uXEAAOA1BBVP+fpKzZpZXQUAAKUCXT8AAMC2uKLiqbQ06bXXsscfe8zaWgAAKOEIKp5KT5eefDJ7fOBAa2sBAKCEo+sHAADYFkEFAADYFkEFAADYVqGCyk033aTjx4/nmp6cnKybbrrp79YEAAAgqZBBZdmyZUpLS8s1PSUlRT/99NPfLgoAAEDy8KmfjRs3usZ/++03HThwwPU5MzNT33zzjapUqVJ01QEAgFLNo6DSqFEjORwOORyOPLt4AgMDNXXq1CIrzpYCAqSlS8+NAwAAr/EoqOzcuVPGGNWsWVOrV69WpUqVXPP8/f0VHh4uX1/fIi/SVnx9pbZtra4CAIBSwaOgUr16dUlSVlaWV4oBAAA4X6HfTPvHH39o2bJlOnjwYK7gMnbs2L9dmG2lp0szZmSPP/SQtbUAAFDCFSqozJw5U48++qgqVqyoypUry+FwuOY5HI6SHVTS0qTBg7PH+/a1tBQAAEq6QgWViRMnatKkSRo5cmRR1wMAAOBSqPeoHDt2TD169CjqWgAAANwUKqj06NFDixcvLupaAAAA3BSq66d27dp65plntGrVKtWvX19+fn5u84cOHVokxQEAgNKtUEFlxowZCgkJUXx8vOLj493mORwOggoAACgShQoqO3fuLOo6AAAAcin0e1RKLadT+uqrc+MAAMBrChVU+vXrl+/8d999t1DFXBbKlJE6d7a6CgAASoVCBZVjx465fU5PT9fmzZt1/PjxPL+sEAAAoDAKFVQWLlyYa1pWVpYGDhyomjVr/u2ibC09XZo7N3u8d29rawEAoIQr1HtU8lyRj48ef/xx/etf/yqqVdpTWpr0wAPZQ1qa1dUAAFCiFVlQkaTt27crIyOjKFcJAABKsUJ1/QwfPtztszFG+/fv16JFi9SnT58iKQwAAKBQQWX9+vVun318fFSpUiW98sorl3wiCAAAoKAKFVSWLl1a1HUAAADk8rde+Hbo0CElJibK4XDoqquuUqVKlYqqLgAAgMLdTHv69Gn169dPkZGRuvHGG9W6dWtFRUXpwQcf1JkzZ4q6RgAAUEoVKqgMHz5c8fHx+vLLL3X8+HEdP35cn3/+ueLj4zVixIiirtFenE7p44+zB16hDwCAVxWq62fBggX6z3/+o7Zt27qmderUSYGBgerZs6emT59eVPXZT5kyUo8eVlcBAECpUKgrKmfOnFFERESu6eHh4XT9AACAIlOooNK8eXONGzdOKSkprmlnz57VhAkT1Lx58yIrzpYyMqRPPskeeLkdAABeVaiunylTpqhjx46qWrWqGjZsKIfDoQ0bNsjpdGrx4sVFXaO9pKZKPXtmj586ZW0tAACUcIUKKvXr19fWrVv1wQcf6H//+5+MMerVq5d69+6twMDAoq4RAACUUoUKKnFxcYqIiNCAAQPcpr/77rs6dOiQRo4cWSTFAQCA0q1Q96i89dZbqlOnTq7pdevW1Ztvvvm3iwIAAJAKGVQOHDigyMjIXNMrVaqk/fv3/+2iAAAApEIGlejoaP3888+5pv/888+Kior620UBAABIhbxHpX///ho2bJjS09N10003SZK+//57PfnkkyX/zbQAAKDYFCqoPPnkkzp69KgGDhyotLQ0SVJAQIBGjhypUaNGFWmBtuPvL82adW4cAAB4TaGCisPh0AsvvKBnnnlGv//+uwIDA3XllVfKWRq++8bPT+rb1+oqAAAoFQoVVHKEhISoWbNmRVULAACAm78VVEqljAzp22+zx2+5xdpaAAAo4QgqnkpNlbp0yR7nFfoAAHhVoR5P9oa4uDg5HA4NGzbM6lIAAIBN2CKorFmzRjNmzFCDBg2sLgUAANiI5UHl1KlT6t27t2bOnKly5cpZXQ4AALARy4PKoEGD1LlzZ918882XbJuamqrk5GS3AQAAlFyW3kw7f/58rVu3TmvWrClQ+7i4OE2YMMHLVZ0T89Qit89Jz3cukvUUp7y2Xdj9AACguFl2RWXPnj167LHH9MEHHyggIKBAy4waNUonTpxwDXv27PFylQAAwEqWXVFJSEjQwYMH1aRJE9e0zMxM/fjjj/r3v/+t1NRU+fr6ui3jdDqtf/utv7/073+fGwcAAF5jWVBp3769Nm3a5DbtgQceUJ06dTRy5MhcIcU2/PykQYOsrgIAgFLBsqBStmxZ1atXz21acHCwKlSokGs6AAAonXgzracyM6Wffsoeb93a2loAACjhbBVUli1bZnUJl5aSIrVrlz3OK/QBAPAqy9+jAgAAcDEEFQAAYFsEFQAAYFsEFQAAYFsEFQAAYFsEFQAAYFu2ejz5suDnJ7344rlxAADgNQQVT/n7S088YXUVAACUCnT9AAAA2+KKiqcyM6V167LHr73W2loAACjhCCqeSkmRrrsue5xX6AMA4FV0/QAAANsiqAAAANsiqAAAANsiqAAAANsiqAAAANsiqAAAANvi8WRP+flJ48adGwcAAF5DUPGUv780frzVVQAAUCrQ9QMAAGyLKyqeysqSfv89ezw21tpaAAAo4Qgqnjp7VqpXL3ucV+gDAOBVdP0AAADbIqgAAADbIqgAAADbIqgAAADbIqgAAADbIqgAAADb4vFkT/n5Sf/3f+fGAQCA1xBUPOXvL730ktVVAABQKtD1AwAAbIsrKp7KypJ2784er1bN2loAACjhCCqeOntWqlEje9yLr9CPeWpRoZZLer5zEVeCgsjrfHEuYAf8bOJyR9cPAACwLYIKAACwLYIKAACwLYIKAACwLYIKAACwLYIKAACwLR5P9lSZMtLAgefGAQCA1/CX1lNOp/TGG1ZXAQBAqUDXDwAAsC2uqHjKGOnw4ezxihWtrQUAgBKOoOKpM2ek8PDscS++Qh8AAND1AwAAbIygAgAAbIugAgAAbIugAgAAbIugAgAAbIugAgAAbIvHkz1VpozUp8+5cQAA4DX8pfWU0ynNnm11FQAAlAp0/QAAANviioqnjMl+O60kBQVZWwsAACUcQcVTZ85IISHZ47xCHwAAr6LrBwAA2BZBBQAA2BZBBQAA2BZBBQAA2BZBBQAA2BZBBQAA2BaPJ3vK11fq3v3cOAAA8BqCiqcCAqRPPrG6CgAASgW6fgAAgG0RVAAAgG0RVDx1+rTkcGQPp09bXQ0AACUaQQUAANiWpUElLi5OzZo1U9myZRUeHq7bb79diYmJVpYEAABsxNKgEh8fr0GDBmnVqlVasmSJMjIy1KFDB52mSwUAAMjix5O/+eYbt8+zZs1SeHi4EhISdOONN1pUFQAAsAtb3aNy4sQJSVL58uUtrgQAANiBbV74ZozR8OHD1apVK9WrVy/PNqmpqUpNTXV9Tk5OLq7yAACABWwTVAYPHqyNGzdq+fLlF20TFxenCRMmFGNV7mKeWiRnRpqm12wqSXp03GKpjH+RrdtO6wGQ+99T0vOdLarEu/L6vXE57mtpOV/eZMefBVsElSFDhuiLL77Qjz/+qKpVq1603ahRozR8+HDX5+TkZEVHRxdHiS6pZfzVr8f4Yt0mAACllaVBxRijIUOGaOHChVq2bJlq1KiRb3un0ymn01lM1QEAAKtZGlQGDRqkefPm6fPPP1fZsmV14MABSVJYWJgCAwOtLA0AANiApU/9TJ8+XSdOnFDbtm0VGRnpGj766CMry8pXYFqKfnv1Tv326p0KTEuxuhwAAEo0y7t+LkdB6amXbgQAAP42W71HBQAA4HwEFQAAYFsEFQAAYFsEFQAAYFsEFQAAYFu2eDPt5STL4dCq6HqucQAA4D0EFQ+l+jnV657nrS4DAIBSga4fAABgWwQVAABgWwQVDwWmpSjh9XuU8Po9vEIfAAAv4x6VQqhwNtnqEgAAKBW4ogIAAGyLoAIAAGyLoAIAAGyLoAIAAGyLoAIAAGyLp348lOVw6NfKV7rGAQCA9xBUPJTq51TXPv+yugwAAEoFun4AAIBtEVQAAIBtEVQ8FJCeouXT+2n59H4KSOcV+gAAeBP3qHjIYaSqyQdd4wAAwHu4ogIAAGyLoAIAAGyLoAIAAGyLoAIAAGyLoAIAAGyLp348ZBzSHxWqucYBAID3EFQ8lOIXoA79p1ldBgAApQJdPwAAwLYIKgAAwLYIKh4KSE/R4rcHavHbA3mFPgAAXsY9Kh5yGOmqI7td4wAAwHsIKoCFYp5alGta0vOdLajk4i6sMa/6CtLGbrxV8+V4Tgvbxlv7dTkcwwsV5Hjlxe77ZQd0/QAAANsiqAAAANsiqAAAANsiqAAAANviZloPGYf0Z2i4axwAAHgPQcVDKX4BavXou1aXAQBAqUDXDwAAsC2CCgAAsC2Cioec6an6fM7j+nzO43Kmp1pdDgAAJRr3qHjIxxg1PLDVNQ4AALyHKyoAAMC2CCoAAMC2CCoAAMC2CCoAAMC2CCoAAMC2eOqnEI4EhlpdAgAApQJBxUNn/QPUZOg8q8sAAKBUoOsHAADYFkEFAADYFkHFQ870VM2f95Tmz3uKV+gDAOBl3KPiIR9jdMOeza5xAADgPVxRAQAAtkVQAQAAtkVQAQAAtkVQAQAAtkVQAQAAtsVTP4Vwxs9pdQkAAJQKBBUPnfUP0DXDF1hdBgAApQJdPwAAwLYIKgAAwLbo+vGQMyNN0xdOliQ92m20Usv4W1wRAAAlF0HFQz5ZWbppx1rXOAAA8B66fgAAgG1ZHlSmTZumGjVqKCAgQE2aNNFPP/1kdUkAAMAmLA0qH330kYYNG6YxY8Zo/fr1at26tTp27Kjdu3dbWRYAALAJS4PKq6++qgcffFD9+/dXbGyspkyZoujoaE2fPt3KsgAAgE1YFlTS0tKUkJCgDh06uE3v0KGDVqxYYVFVAADATix76ufw4cPKzMxURESE2/SIiAgdOHAgz2VSU1OVmprq+nzixAlJUnJysldqzEo9k2taZlqKcraWmXpGWebye/LHW8erNMvrZ6Ugx7mwyxWnC2vMq76CtLEbb+2X3c5pXvUUleL83VuYY1+cx72wx9lu/1aK6+c3Z53GmEs3NhbZu3evkWRWrFjhNn3ixInm6quvznOZcePGGUkMDAwMDAwMJWDYs2fPJfOCZVdUKlasKF9f31xXTw4ePJjrKkuOUaNGafjw4a7PWVlZOnr0qCpUqCCHw+HVeq2SnJys6Oho7dmzR6GhoVaXU+pxPuyF82EvnA97sfP5MMbo5MmTioqKumRby4KKv7+/mjRpoiVLlqhbt26u6UuWLFHXrl3zXMbpdMrpdP/m4iuuuMKbZdpGaGio7X7QSjPOh71wPuyF82Evdj0fYWFhBWpn6Ztphw8frvvuu09NmzZV8+bNNWPGDO3evVuPPPKIlWUBAACbsDSo3HXXXTpy5IieffZZ7d+/X/Xq1dN///tfVa9e3cqyAACATVj+XT8DBw7UwIEDrS7DtpxOp8aNG5erywvW4HzYC+fDXjgf9lJSzofDmII8GwQAAFD8LP+uHwAAgIshqAAAANsiqAAAANsiqAAAANsiqBSz6dOnq0GDBq4X8DRv3lxff/21a74xRuPHj1dUVJQCAwPVtm1bbdmyxW0dqampGjJkiCpWrKjg4GDddttt+vPPP4t7V0qkuLg4ORwODRs2zDWNc1J8xo8fL4fD4TZUrlzZNZ9zUfz27t2re++9VxUqVFBQUJAaNWqkhIQE13zOSfGJiYnJ9e/D4XBo0KBBkkrwufhbX9gDj33xxRdm0aJFJjEx0SQmJprRo0cbPz8/s3nzZmOMMc8//7wpW7asWbBggdm0aZO56667TGRkpElOTnat45FHHjFVqlQxS5YsMevWrTPt2rUzDRs2NBkZGVbtVomwevVqExMTYxo0aGAee+wx13TOSfEZN26cqVu3rtm/f79rOHjwoGs+56J4HT161FSvXt307dvX/PLLL2bnzp3mu+++M9u2bXO14ZwUn4MHD7r921iyZImRZJYuXWqMKbnngqBiA+XKlTNvv/22ycrKMpUrVzbPP/+8a15KSooJCwszb775pjHGmOPHjxs/Pz8zf/58V5u9e/caHx8f88033xR77SXFyZMnzZVXXmmWLFli2rRp4woqnJPiNW7cONOwYcM853Euit/IkSNNq1atLjqfc2Ktxx57zNSqVctkZWWV6HNB14+FMjMzNX/+fJ0+fVrNmzfXzp07deDAAXXo0MHVxul0qk2bNlqxYoUkKSEhQenp6W5toqKiVK9ePVcbeG7QoEHq3Lmzbr75ZrfpnJPit3XrVkVFRalGjRrq1auXduzYIYlzYYUvvvhCTZs2VY8ePRQeHq7GjRtr5syZrvmcE+ukpaXpgw8+UL9+/eRwOEr0uSCoWGDTpk0KCQmR0+nUI488ooULF+qaa65xfZP0hd8eHRER4Zp34MAB+fv7q1y5chdtA8/Mnz9f69atU1xcXK55nJPidf311+u9997Tt99+q5kzZ+rAgQNq0aKFjhw5wrmwwI4dOzR9+nRdeeWV+vbbb/XII49o6NCheu+99yTx78NKn332mY4fP66+fftKKtnnwvJX6JdGV199tTZs2KDjx49rwYIF6tOnj+Lj413zHQ6HW3tjTK5pFypIG+S2Z88ePfbYY1q8eLECAgIu2o5zUjw6duzoGq9fv76aN2+uWrVqac6cObrhhhskcS6KU1ZWlpo2barJkydLkho3bqwtW7Zo+vTpuv/++13tOCfF75133lHHjh0VFRXlNr0knguuqFjA399ftWvXVtOmTRUXF6eGDRvqtddecz3dcGGyPXjwoCslV65cWWlpaTp27NhF26DgEhISdPDgQTVp0kRlypRRmTJlFB8fr9dff11lypRxHVPOiTWCg4NVv359bd26lX8fFoiMjNQ111zjNi02Nla7d++WJM6JRXbt2qXvvvtO/fv3d00ryeeCoGIDxhilpqaqRo0aqly5spYsWeKal5aWpvj4eLVo0UKS1KRJE/n5+bm12b9/vzZv3uxqg4Jr3769Nm3apA0bNriGpk2bqnfv3tqwYYNq1qzJObFQamqqfv/9d0VGRvLvwwItW7ZUYmKi27Q//vjD9Q33nBNrzJo1S+Hh4ercubNrWok+FxbdxFtqjRo1yvz4449m586dZuPGjWb06NHGx8fHLF682BiT/XhZWFiY+fTTT82mTZvM3XffnefjZVWrVjXfffedWbdunbnpppts/3jZ5eT8p36M4ZwUpxEjRphly5aZHTt2mFWrVpkuXbqYsmXLmqSkJGMM56K4rV692pQpU8ZMmjTJbN261cydO9cEBQWZDz74wNWGc1K8MjMzTbVq1czIkSNzzSup54KgUsz69etnqlevbvz9/U2lSpVM+/btXSHFmOzH/caNG2cqV65snE6nufHGG82mTZvc1nH27FkzePBgU758eRMYGGi6dOlidu/eXdy7UmJdGFQ4J8Un570Pfn5+Jioqytxxxx1my5Ytrvmci+L35Zdfmnr16hmn02nq1KljZsyY4Tafc1K8vv32WyPJJCYm5ppXUs+FwxhjrL6qAwAAkBfuUQEAALZFUAEAALZFUAEAALZFUAEAALZFUAEAALZFUAEAALZFUAEAALZFUAFQIrVt21bDhg2zugxJ0rJly+RwOHT8+HGrSwEuOwQVAChCdgpIQElAUAEAALZFUAEuY//5z39Uv359BQYGqkKFCrr55pt1+vRp1/xZs2YpNjZWAQEBqlOnjqZNm+a2/OrVq9W4cWMFBASoadOmWrhwoRwOhzZs2CBJmj17tq644gq3ZT777DM5HA63aV9++aWaNGmigIAA1axZUxMmTFBGRoZrvsPh0Ntvv61u3bopKChIV155pb744gu3dWzZskWdO3dWaGioypYtq9atW2v79u0F3pdLSUtL05NPPqkqVaooODhY119/vZYtW+aan7Ov3377rWJjYxUSEqJ//vOf2r9/v6tNRkaGhg4dqiuuuEIVKlTQyJEj1adPH91+++2SpL59+yo+Pl6vvfaaHA6HHA6HkpKSXMsnJCSoadOmCgoKUosWLXJ9MzGAPFj9ZUMACmffvn2mTJky5tVXX3V9G/cbb7xhTp48aYwxZsaMGSYyMtIsWLDA7NixwyxYsMCUL1/ezJ492xhjzKlTp0ylSpXMXXfdZTZv3my+/PJLU7NmTSPJrF+/3hhjzKxZs0xYWJjbdhcuXGjO/9XxzTffmNDQUDN79myzfft2s3jxYhMTE2PGjx/vaiPJVK1a1cybN89s3brVDB061ISEhJgjR44YY4z5888/Tfny5c0dd9xh1qxZYxITE827775r/ve//xVoX/Jy4ZdL3nPPPaZFixbmxx9/NNu2bTMvvfSScTqd5o8//nDtq5+fn7n55pvNmjVrTEJCgomNjTX33HOPax0TJ0405cuXN59++qn5/fffzSOPPGJCQ0NN165djTHGHD9+3DRv3twMGDDA7N+/3+zfv99kZGSYpUuXGknm+uuvN8uWLTNbtmwxrVu3Ni1atPDgjAOlE0EFuEwlJCQYSSYpKSnP+dHR0WbevHlu05577jnTvHlzY4wxb731lilfvrw5ffq0a/706dM9DiqtW7c2kydPdmvz/vvvm8jISNdnSebpp592fT516pRxOBzm66+/NsYYM2rUKFOjRg2TlpZWqH3Jy/lBZdu2bcbhcJi9e/e6tWnfvr0ZNWqUa18lmW3btrnmv/HGGyYiIsL1OSIiwrz00kuuzxkZGaZatWquoHLhdnPkBJXvvvvONW3RokVGkjl79uxF9wGAMWUsupAD4G9q2LCh2rdvr/r16+uWW25Rhw4d1L17d5UrV06HDh3Snj179OCDD2rAgAGuZTIyMhQWFiZJ+v3339WwYUMFBQW55jdv3tzjOhISErRmzRpNmjTJNS0zM1MpKSk6c+aMa/0NGjRwzQ8ODlbZsmV18OBBSdKGDRvUunVr+fn55Vp/QfblUtatWydjjK666iq36ampqapQoYLrc1BQkGrVquX6HBkZ6arxxIkT+uuvv3Tddde55vv6+qpJkybKysoqUB3nH4PIyEhJ0sGDB1WtWrUCLQ+URgQV4DLl6+urJUuWaMWKFVq8eLGmTp2qMWPG6JdffnGFg5kzZ+r666/PtZwkGWMuuQ0fH59c7dLT090+Z2VlacKECbrjjjtyLR8QEOAavzCEOBwO1x/4wMDAi9aQ0ya/fbmUrKws+fr6KiEhIdcyISEh+dZ44f5feH9OQY5jXuvPWU9BQw5QWhFUgMuYw+FQy5Yt1bJlS40dO1bVq1fXwoULNXz4cFWpUkU7duxQ796981z2mmuu0fvvv6+zZ8+6gsKqVavc2lSqVEknT57U6dOnFRwcLEmuG21zXHvttUpMTFTt2rULvR8NGjTQnDlzlJ6enissREREXHJfLqVx48bKzMzUwYMH1bp160KtIywsTBEREVq9erVrHZmZmVq/fr0aNWrkaufv76/MzMxCbQNAbgQV4DL1yy+/6Pvvv1eHDh0UHh6uX375RYcOHVJsbKwkafz48Ro6dKhCQ0PVsWNHpaamau3atTp27JiGDx+ue+65R2PGjNGDDz6op59+WklJSXr55ZfdtnH99dcrKChIo0eP1pAhQ7R69WrNnj3brc3YsWPVpUsXRUdHq0ePHvLx8dHGjRu1adMmTZw4sUD7MnjwYE2dOlW9evXSqFGjFBYWplWrVum6667T1Vdffcl9uZSrrrpKvXv31v33369XXnlFjRs31uHDh/XDDz+ofv366tSpU4HqHDJkiOLi4lS7dm3VqVNHU6dO1bFjx9yussTExOiXX35RUlKSQkJCVL58+QKtG8BFWHqHDIBC++2338wtt9xiKlWqZJxOp7nqqqvM1KlT3drMnTvXNGrUyPj7+5ty5cqZG2+80Xz66aeu+StXrjQNGzY0/v7+plGjRmbBggVuN9Mak33zbO3atU1AQIDp0qWLmTFjhrnwV8c333xjWrRoYQIDA01oaKi57rrrzIwZM1zzJZmFCxe6LRMWFmZmzZrl+vzrr7+aDh06mKCgIFO2bFnTunVrs3379gLvy4UuvKk1LS3NjB071sTExBg/Pz9TuXJl061bN7Nx40ZjTMFuHE5PTzeDBw82oaGhply5cmbkyJGmR48eplevXq42iYmJ5oYbbjCBgYFGktm5c6frZtpjx4652q1fv941H8DFOYzxoIMVQImWlJSkGjVq5OrOQN6ysrIUGxurnj176rnnnrO6HKBEousHAApo165dWrx4sdq0aaPU1FT9+9//1s6dO3XPPfdYXRpQYvFmWgAoIB8fH82ePVvNmjVTy5YttWnTJn333Xeu+4IAFD26fgAAgG1xRQUAANgWQQUAANgWQQUAANgWQQUAANgWQQUAANgWQQUAANgWQQUAANgWQQUAANgWQQUAANjW/wNc5c0Qnw02rQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Now, we want to visualize the length distribution of the MMSeqs2 hits to filter by length. \n",
    "query_id = 'UZE92782.1'\n",
    "# query_id = 'UZE92041.1'\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "hit_lengths = mmseqs_df[mmseqs_df.query_id.str.match(query_id)].target_seq.apply(lambda s : len(s.replace('*', '')))\n",
    "_, bins, _ = ax.hist(hit_lengths, bins=100)\n",
    "print(f'Mean length of hits for {query_id}:', np.mean(hit_lengths))\n",
    "\n",
    "ymin, ymax = ax.get_ylim()\n",
    "query_length = mmseqs_df[mmseqs_df.query_id.str.match(query_id)].query_aa_length.iloc[0] \n",
    "ax.axvline(int(query_length), ymin=ymin, ymax=ymax, c='red', ls='--')\n",
    "\n",
    "ax.set_title(f'Length distribution for {query_id} hits')\n",
    "ax.set_ylabel('count')\n",
    "ax.set_xlabel('sequence length')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Based on the sequence length plot, we can select reasonable boundaries for the subsequent BLAST search. \n",
    "# There seems to be more hits for the sample UZE92041.1, so we will focus on that. \n",
    "mmseqs_df = mmseqs_df[mmseqs_df.query_id.str.match('UZE92782.1')]\n",
    "mmseqs_df = mmseqs_df[mmseqs_df.target_seq.apply(len) < 400]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequences remaining after filter: 49\n"
     ]
    }
   ],
   "source": [
    "print('Sequences remaining after filter:', len(mmseqs_df))\n",
    "dataframe_to_fasta(mmseqs_df.rename(columns={'target_seq':'seq'}), f'{DATA_DIR}{genome_id}/blast_target.faa')\n",
    "\n",
    "blast_query_df = dataframe_from_fasta(f'{DATA_DIR}{genome_id}/query.faa')\n",
    "blast_query_df = blast_query_df[blast_query_df['id'].str.match(query_id)]\n",
    "dataframe_to_fasta(blast_query_df, f'{DATA_DIR}{genome_id}/blast_query.faa')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aligning homologs against extended sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install the BLAST command-line tool for Linux, and unzip the file in your home directory. \n",
    "curl https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.15.0+-x64-linux.tar.gz  -o /home/prichter/\n",
    "tar /home/prichter/zxvpf ncbi-blast-2.15.0+-x64-linux.tar.gz "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the path to the blast alignment command. \n",
    "blastp_cmd = '/home/prichter/ncbi-blast-2.15.0+/bin/blastp'\n",
    "# Define a custom TSV output format for the BLAST alignment. \n",
    "outfmt = '6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qseq sseq'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "! '{blastp_cmd}' -subject '{DATA_DIR}{genome_id}/blast_target.faa' -query '{DATA_DIR}{genome_id}/blast_query.faa' -outfmt '{outfmt}' -out '{DATA_DIR}{genome_id}/blast_results.m8'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>percentage_identical</th>\n",
       "      <th>align_length</th>\n",
       "      <th>num_mismatches</th>\n",
       "      <th>num_gap_openings</th>\n",
       "      <th>query_align_start</th>\n",
       "      <th>query_align_stop</th>\n",
       "      <th>target_align_start</th>\n",
       "      <th>target_align_stop</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>...</th>\n",
       "      <th>query_scaffold_id</th>\n",
       "      <th>query_feature</th>\n",
       "      <th>query_nt_start</th>\n",
       "      <th>query_nt_stop</th>\n",
       "      <th>query_reverse</th>\n",
       "      <th>query_frame</th>\n",
       "      <th>query_gene_id</th>\n",
       "      <th>query_nt_ext</th>\n",
       "      <th>query_aa_length</th>\n",
       "      <th>query_u_codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>33.526</td>\n",
       "      <td>173</td>\n",
       "      <td>113</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>208</td>\n",
       "      <td>116</td>\n",
       "      <td>286</td>\n",
       "      <td>1.900000e-34</td>\n",
       "      <td>118.0</td>\n",
       "      <td>...</td>\n",
       "      <td>CP061209.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>182508</td>\n",
       "      <td>183327</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>24</td>\n",
       "      <td>273</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23.626</td>\n",
       "      <td>182</td>\n",
       "      <td>131</td>\n",
       "      <td>3</td>\n",
       "      <td>62</td>\n",
       "      <td>241</td>\n",
       "      <td>102</td>\n",
       "      <td>277</td>\n",
       "      <td>4.110000e-18</td>\n",
       "      <td>73.6</td>\n",
       "      <td>...</td>\n",
       "      <td>CP061209.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>182508</td>\n",
       "      <td>183327</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>24</td>\n",
       "      <td>273</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24.503</td>\n",
       "      <td>151</td>\n",
       "      <td>110</td>\n",
       "      <td>2</td>\n",
       "      <td>62</td>\n",
       "      <td>210</td>\n",
       "      <td>123</td>\n",
       "      <td>271</td>\n",
       "      <td>5.230000e-17</td>\n",
       "      <td>70.1</td>\n",
       "      <td>...</td>\n",
       "      <td>CP061209.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>182508</td>\n",
       "      <td>183327</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>24</td>\n",
       "      <td>273</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26.347</td>\n",
       "      <td>167</td>\n",
       "      <td>108</td>\n",
       "      <td>4</td>\n",
       "      <td>62</td>\n",
       "      <td>217</td>\n",
       "      <td>114</td>\n",
       "      <td>276</td>\n",
       "      <td>9.940000e-17</td>\n",
       "      <td>69.3</td>\n",
       "      <td>...</td>\n",
       "      <td>CP061209.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>182508</td>\n",
       "      <td>183327</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>24</td>\n",
       "      <td>273</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22.013</td>\n",
       "      <td>159</td>\n",
       "      <td>115</td>\n",
       "      <td>2</td>\n",
       "      <td>31</td>\n",
       "      <td>189</td>\n",
       "      <td>106</td>\n",
       "      <td>255</td>\n",
       "      <td>1.270000e-16</td>\n",
       "      <td>69.3</td>\n",
       "      <td>...</td>\n",
       "      <td>CP061209.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>182508</td>\n",
       "      <td>183327</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>24</td>\n",
       "      <td>273</td>\n",
       "      <td>TAA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   percentage_identical  align_length  num_mismatches  num_gap_openings  \\\n",
       "0                33.526           173             113                 1   \n",
       "1                23.626           182             131                 3   \n",
       "2                24.503           151             110                 2   \n",
       "3                26.347           167             108                 4   \n",
       "4                22.013           159             115                 2   \n",
       "\n",
       "   query_align_start  query_align_stop  target_align_start  target_align_stop  \\\n",
       "0                 36               208                 116                286   \n",
       "1                 62               241                 102                277   \n",
       "2                 62               210                 123                271   \n",
       "3                 62               217                 114                276   \n",
       "4                 31               189                 106                255   \n",
       "\n",
       "        e_value  bit_score  ... query_scaffold_id query_feature  \\\n",
       "0  1.900000e-34      118.0  ...        CP061209.1           CDS   \n",
       "1  4.110000e-18       73.6  ...        CP061209.1           CDS   \n",
       "2  5.230000e-17       70.1  ...        CP061209.1           CDS   \n",
       "3  9.940000e-17       69.3  ...        CP061209.1           CDS   \n",
       "4  1.270000e-16       69.3  ...        CP061209.1           CDS   \n",
       "\n",
       "   query_nt_start  query_nt_stop  query_reverse  query_frame  query_gene_id  \\\n",
       "0          182508         183327           True            0           None   \n",
       "1          182508         183327           True            0           None   \n",
       "2          182508         183327           True            0           None   \n",
       "3          182508         183327           True            0           None   \n",
       "4          182508         183327           True            0           None   \n",
       "\n",
       "   query_nt_ext  query_aa_length  query_u_codon  \n",
       "0            24              273            TAA  \n",
       "1            24              273            TAA  \n",
       "2            24              273            TAA  \n",
       "3            24              273            TAA  \n",
       "4            24              273            TAA  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blastp_df = dataframe_from_m8(f'{DATA_DIR}{genome_id}/blast_results.m8')\n",
    "blastp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maximum sequence identity: 33.526\n",
      "Minimum sequence identity: 19.512\n",
      "Mean sequence identity: 23.971540000000005\n"
     ]
    }
   ],
   "source": [
    "print('Maximum sequence identity:', max(blastp_df.percentage_identical))\n",
    "print('Minimum sequence identity:', min(blastp_df.percentage_identical))\n",
    "print('Mean sequence identity:', blastp_df.percentage_identical.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JAFRMD010000040.1_2\n",
      "CACXCN010000042.1_15\n",
      "CAISVC010000334.1_1\n",
      "SLMV01000216.1_7\n",
      "CAJORE010000133.1_2\n",
      "JAFQMU010000191.1_11\n",
      "JAFTCE010000087.1_14\n",
      "CACWQW010000002.1_20\n",
      "SKJJ01000116.1_9\n",
      "JADFVM010000040.1_20\n",
      "NZ_AUBP01000002.1_72\n",
      "SLGS01000155.1_2\n",
      "WQRV01000016.1_6\n",
      "JAAYOY010000184.1_5\n",
      "JAABTT010000029.1_23\n",
      "JACRNJ010000069.1_9\n",
      "JACRCM010000168.1_26\n",
      "MHDO01000099.1_5\n",
      "JACQXH010000220.1_6\n",
      "VGSO01000021.1_8\n",
      "JACNJG010000027.1_7\n",
      "NZ_ARQF01000020.1_864\n",
      "JACQQJ010000126.1_9\n",
      "JACRPW010000006.1_35\n",
      "QKEJ01000196.1_5\n",
      "QKEJ01000196.1_5\n",
      "MGRL01000091.1_9\n",
      "JAAYGN010000044.1_4\n",
      "DTGR01000225.1_7\n",
      "DNJG01000034.1_8\n",
      "MNZR01000272.1_3\n",
      "CABKMN010000070.1_11\n",
      "NZ_AUCY01000041.1_19\n",
      "LR026961.1_3008\n",
      "NZ_AULM01000012.1_90\n",
      "NZ_JYNO01000004.1_377\n",
      "DXFI01000243.1_4\n",
      "JAHJUF010000009.1_1\n",
      "DBMO01000023.1_15\n",
      "JACNJD010000247.1_3\n",
      "NZ_JACHGO010000006.1_23\n",
      "LUZZ01000011.1_8\n",
      "PQXE01000019.1_20\n",
      "DSCZ01000118.1_2\n",
      "WGDC01000016.1_26\n",
      "DREN01000450.1_5\n",
      "CAJUHZ010000014.1_8\n",
      "PQXD01000004.1_69\n",
      "JAGYNZ010000008.1_28\n",
      "DTYO01000048.1_3\n"
     ]
    }
   ],
   "source": [
    "for target_id in blastp_df.target_id:\n",
    "    print(target_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "selenobot",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
